<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Importing From Other Sources</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Importing From Other Sources</h1>



<p>MOCHA is intended for use after clustering and cell labeling, since
all functions of MOCHA operate on per-sample basis within each cell
population. MOCHA can be used with an ArchR Project as input, as well as
with a generic input. This tutorial has two sections: demonstrating
MOCHA with generic inputs - <a href="#signac">Importing from Signac</a>
and <a href="#snapatac">Importing from SnapATAC</a>.</p>
<div id="signac" class="section level2">
<h2>Importing from Signac</h2>
<p>This vignette will follow the <a href="https://stuartlab.org/signac/articles/pbmc_multiomic.html">Signac
tutorial</a> to generate an Signac object and label cell types within
it. If you already have a Signac object with cell types labelled, skip
to section <a href="#extract-frags-signac">Extract Fragments from
Signac</a></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Signac)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratDisk)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnsDb.Hsapiens.v86)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenome.Hsapiens.UCSC.hg38)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span></code></pre></div>
<div id="signac-tutorial-through-clustering" class="section level3">
<h3>Signac Tutorial Through Clustering</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download files</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">&#39;wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5&#39;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">&#39;wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz&#39;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">&#39;wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi&#39;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in ATAC and RNA data</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">Read10X_h5</span>(<span class="st">&quot;pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5&quot;</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>fragpath <span class="ot">&lt;-</span> <span class="st">&quot;pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a Seurat object containing the RNA adata</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> counts<span class="sc">$</span><span class="st">`</span><span class="at">Gene Expression</span><span class="st">`</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">&quot;RNA&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># create ATAC assay and add it to the object</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>pbmc[[<span class="st">&quot;ATAC&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateChromatinAssay</span>(</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> counts<span class="sc">$</span>Peaks,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">sep =</span> <span class="fu">c</span>(<span class="st">&quot;:&quot;</span>, <span class="st">&quot;-&quot;</span>),</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">fragments =</span> fragpath,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotation =</span> annotation</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;ATAC&quot;</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">NucleosomeSignal</span>(pbmc)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">TSSEnrichment</span>(pbmc)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> pbmc,</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">subset =</span> nCount_ATAC <span class="sc">&lt;</span> <span class="dv">100000</span> <span class="sc">&amp;</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    nCount_RNA <span class="sc">&lt;</span> <span class="dv">25000</span> <span class="sc">&amp;</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    nCount_ATAC <span class="sc">&gt;</span> <span class="dv">1000</span> <span class="sc">&amp;</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>    nCount_RNA <span class="sc">&gt;</span> <span class="dv">1000</span> <span class="sc">&amp;</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>    nucleosome_signal <span class="sc">&lt;</span> <span class="dv">2</span> <span class="sc">&amp;</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>    TSS.enrichment <span class="sc">&gt;</span> <span class="dv">1</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform RNA </span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;RNA&quot;</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(pbmc)</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(pbmc)</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>pbmc[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(pbmc, <span class="at">pattern =</span> <span class="st">&quot;^MT-&quot;</span>)</span></code></pre></div>
<p>Label cell types:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load PBMC reference</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">&#39;wget https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat&#39;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;pbmc_multimodal.h5seurat&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;SCT&quot;</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># transfer cell type labels from reference to query</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>transfer_anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference =</span> reference,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">query =</span> pbmc,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">normalization.method =</span> <span class="st">&quot;SCT&quot;</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference.reduction =</span> <span class="st">&quot;spca&quot;</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">recompute.residuals =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">anchorset =</span> transfer_anchors, </span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> reference<span class="sc">$</span>celltype.l2,</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">weight.reduction =</span> pbmc[[<span class="st">&#39;pca&#39;</span>]],</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> predictions</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set the cell identities to the cell type predictions</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;predicted.id&quot;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># set a reasonable order for cell types to be displayed when plotting</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(pbmc) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;CD4 Naive&quot;</span>, <span class="st">&quot;CD4 TCM&quot;</span>, <span class="st">&quot;CD4 CTL&quot;</span>, <span class="st">&quot;CD4 TEM&quot;</span>, <span class="st">&quot;CD4 Proliferating&quot;</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;CD8 Naive&quot;</span>, <span class="st">&quot;dnT&quot;</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;CD8 TEM&quot;</span>, <span class="st">&quot;CD8 TCM&quot;</span>, <span class="st">&quot;CD8 Proliferating&quot;</span>, <span class="st">&quot;MAIT&quot;</span>, <span class="st">&quot;NK&quot;</span>, <span class="st">&quot;NK_CD56bright&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;NK Proliferating&quot;</span>, <span class="st">&quot;gdT&quot;</span>,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;Treg&quot;</span>, <span class="st">&quot;B naive&quot;</span>, <span class="st">&quot;B intermediate&quot;</span>, <span class="st">&quot;B memory&quot;</span>, <span class="st">&quot;Plasmablast&quot;</span>,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;CD14 Mono&quot;</span>, <span class="st">&quot;CD16 Mono&quot;</span>,</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;cDC1&quot;</span>, <span class="st">&quot;cDC2&quot;</span>, <span class="st">&quot;pDC&quot;</span>, <span class="st">&quot;HSPC&quot;</span>, <span class="st">&quot;Eryth&quot;</span>, <span class="st">&quot;ASDC&quot;</span>, <span class="st">&quot;ILC&quot;</span>, <span class="st">&quot;Platelet&quot;</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pbmc, <span class="st">&#39;pbmc_Signac_tutorial.rds&#39;</span>)</span></code></pre></div>
</div>
<div id="extract-frags-signac" class="section level3">
<h3>Extract Fragments from Signac Object</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&#39;pbmc_Signac_tutorial.rds&#39;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">&#39;ATAC&#39;</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>fragObj <span class="ot">&lt;-</span> <span class="fu">Fragments</span>(pbmc)</span></code></pre></div>
<p>Signac’s training dataset only has one sample, so to simulate
multiple samples, we will duplicate the data here. <strong><em>This
should not be necessary with a large dataset.</em></strong></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>full_pbmc <span class="ot">&lt;-</span> <span class="fu">merge</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  pbmc, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(pbmc, pbmc), </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">add.cell.ids =</span> <span class="fu">c</span>(<span class="st">&#39;Sample1&#39;</span>, <span class="st">&#39;Sample2&#39;</span>, <span class="st">&#39;Sample3&#39;</span>),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">&#39;DuplicateData&#39;</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>) </span></code></pre></div>
<p>For a larger dataset, you would interact over the fragObj list and
extract each fragments.tsv.gz file.</p>
<p>Instead, we’re just duplicating data for the sake of this
tutorial.</p>
<p>More importantly, you do need to modify the barcode in the fragment
file so it matches the barcodes in Seurat’s metadata.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>fragList <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">mclapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="cf">function</span>(x){</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    frags <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">GetFragmentData</span>(fragObj[[<span class="dv">1</span>]]))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(frags) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;chr&#39;</span>, <span class="st">&#39;start&#39;</span>, <span class="st">&#39;end&#39;</span>, <span class="st">&#39;barcode&#39;</span>, <span class="st">&#39;val&#39;</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    frags<span class="sc">$</span>barcode <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;Sample&quot;</span>,x,<span class="st">&quot;_&quot;</span>, frags<span class="sc">$</span>barcode, <span class="at">sep =</span><span class="st">&#39;&#39;</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    frags</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>}, <span class="at">mc.cores =</span> <span class="dv">3</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(fragList) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;Sample1&#39;</span>, <span class="st">&#39;Sample2&#39;</span>, <span class="st">&#39;Sample3&#39;</span>)</span></code></pre></div>
</div>
<div id="format-metadata-and-fragments-for-mocha" class="section level3">
<h3>Format metadata and fragments for MOCHA</h3>
<p>Generate your sample/cell type list by finding all combinations of
Samples and Cell Populations</p>
<p>Here we must also rename the GRangesList to have names in the format
<code>CellPopulation#Sample</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>celltype_sample_list <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">expand.grid</span>(<span class="fu">unique</span>(pbmc<span class="sc">@</span>meta.data<span class="sc">$</span>predicted.id), <span class="fu">names</span>(fragList)), <span class="dv">1</span>, paste, <span class="at">collapse =</span> <span class="st">&quot;#&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract metadata, add the Sample column, as well as CellBarcode. </span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>fullMeta <span class="ot">&lt;-</span> full_pbmc<span class="sc">@</span>meta.data</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>fullMeta<span class="sc">$</span>Sample <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">&quot;_.*&quot;</span>,<span class="st">&quot;&quot;</span>, <span class="fu">rownames</span>(full_pbmc<span class="sc">@</span>meta.data))</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>fullMeta<span class="sc">$</span>CellBarcode <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">&quot;.*_&quot;</span>,<span class="st">&quot;&quot;</span>, <span class="fu">rownames</span>(full_pbmc<span class="sc">@</span>meta.data))</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Change the CellPopulation_Sample format to CellPopulation#Sample for MOCHA</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(fullMeta) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;_&quot;</span>,<span class="st">&quot;#&quot;</span>, <span class="fu">rownames</span>(fullMeta))</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>CellType_GRanges <span class="ot">&lt;-</span> pbapply<span class="sc">::</span><span class="fu">pblapply</span>(<span class="at">cl =</span> <span class="dv">30</span>, celltype_sample_list, <span class="cf">function</span>(x){</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  celltype <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;#.*&quot;</span>,<span class="st">&quot;&quot;</span>, x)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  sample <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;.*#&quot;</span>,<span class="st">&quot;&quot;</span>, x)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  sortedMeta <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(fullMeta, predicted.id <span class="sc">==</span> celltype, Sample <span class="sc">==</span> sample)</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  sortedFrags <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(fragList[[sample]], barcode <span class="sc">%in%</span> <span class="fu">rownames</span>(sortedMeta))</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">makeGRangesFromDataFrame</span>(sortedFrags, <span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(CellType_GRanges) <span class="ot">&lt;-</span> celltype_sample_list</span></code></pre></div>
<p>Calculate the study signal (the median number of fragments per
cell)</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>avg_reads <span class="ot">&lt;-</span> <span class="fu">lapply</span>(fragList, <span class="cf">function</span>(x){</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                filtFrag <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="fu">as.data.frame</span>(x), barcode  <span class="sc">%in%</span> <span class="fu">rownames</span>(fullMeta))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                <span class="fu">as.vector</span>(<span class="fu">table</span>(filtFrag<span class="sc">$</span>barcode))</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>studySignal <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">unlist</span>(avg_reads))</span></code></pre></div>
</div>
<div id="call-open-tiles-with-mocha" class="section level3">
<h3>Call Open Tiles with MOCHA</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Our blacklist comes included with Signac</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>blackList <span class="ot">&lt;-</span> blacklist_hg38_unified</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Call Open Tiles</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>tileResults <span class="ot">&lt;-</span> MOCHA<span class="sc">::</span><span class="fu">callOpenTiles</span>(<span class="at">ATACFragments =</span> CellType_GRanges,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                             <span class="at">cellColData =</span> fullMeta, </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>                             <span class="at">blackList =</span> blackList, </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>                             <span class="at">genome =</span> <span class="st">&#39;BSgenome.Hsapiens.UCSC.hg38&#39;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                             <span class="at">cellPopLabel =</span> <span class="st">&#39;predicted.id&#39;</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                             <span class="at">cellPopulations =</span> fullMeta<span class="sc">$</span>predicted.id,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>                             <span class="at">studySignal =</span> studySignal,</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>                             <span class="at">cellCol =</span> <span class="st">&#39;barcode&#39;</span>,</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>                             <span class="at">TxDb =</span> <span class="st">&quot;TxDb.Hsapiens.UCSC.hg38.refGene&quot;</span>,</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>                             <span class="at">Org =</span> <span class="st">&quot;org.Hs.eg.db&quot;</span>, </span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>                            <span class="at">outDir =</span> <span class="fu">paste</span>(<span class="fu">getwd</span>(),<span class="st">&#39;/MOCHA_Out&#39;</span>, <span class="at">sep =</span> <span class="st">&#39;&#39;</span>), <span class="at">numCores=</span> <span class="dv">35</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>TSAM <span class="ot">&lt;-</span> MOCHA<span class="sc">::</span><span class="fu">getSampleTileMatrix</span>(tileResults, <span class="at">threshold =</span> <span class="fl">0.2</span>, <span class="at">numCores =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
</div>
<div id="snapatac" class="section level2">
<h2>Importing from SnapATAC</h2>
<p>In this example, we follow the <a href="https://github.com/r3fang/SnapATAC/tree/master/examples/10X_PBMC_15K#data_download">SnapATAC
10X PBMC tutorial</a> through the clustering step before extracting the
fragments and cell metadata necessary for MOCHA.</p>
<p>If you have a fully-formed Snap file for your analysis with
clustering results and sample information added to the metadata, skip to
section <a href="#format-metadata">Formatting Snap File Metadata</a>. If
following the vignette, we STRONGLY recommending installing this patched
version of SnapATAC from <a href="https://github.com/imran-aifi/SnapATAC">this repository</a> with
<code>devtools::install_github(&quot;imran-aifi/SnapATAC&quot;)</code>.</p>
<div id="snapatac-tutorial-through-clustering" class="section level3">
<h3>SnapATAC Tutorial through Clustering</h3>
<p>Download the all the SnapATAC Tutorial Data (linked above) to your
working directory, and load it:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CLI tool for installing from Google Drive share links</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install gdown</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1YiYd_Ydes3tqsJGpNuqQquUOoVj2EEjE/view?usp=share_link</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1YiYd_Ydes3tqsJGpNuqQquUOoVj2EEjE</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1NvGn4M2_HD06PL5Nj2if5xO-uVA0y8Q5/view?usp=share_link</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1NvGn4M2_HD06PL5Nj2if5xO-uVA0y8Q5</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1LUOqsXoQN6lVx-4RNlgH90e5oXQ0y9Bd/view?usp=share_link</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1LUOqsXoQN6lVx-4RNlgH90e5oXQ0y9Bd</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1oMJ6wFsfS-q-sY_yaLEtnYebM7RrBix6/view?usp=share_link</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1oMJ6wFsfS-q-sY_yaLEtnYebM7RrBix6</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1SEFZ5CJgcmoAkmo4_1kCMYS60YOFP379/view?usp=share_link</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1SEFZ5CJgcmoAkmo4_1kCMYS60YOFP379</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1RlBvTCqz6mhaTAkfYiCdeojD-U2mN2wp/view?usp=share_link</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1RlBvTCqz6mhaTAkfYiCdeojD-U2mN2wp</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SnapATAC);</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>snap.files <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;atac_pbmc_5k_nextgem.snap&quot;</span>, </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;atac_pbmc_10k_nextgem.snap&quot;</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>sample.names <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;PBMC 5K&quot;</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;PBMC 10K&quot;</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>barcode.files <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;atac_pbmc_5k_nextgem_singlecell.csv&quot;</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;atac_pbmc_10k_nextgem_singlecell.csv&quot;</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>x.sp.ls <span class="ot">=</span> <span class="fu">lapply</span>(<span class="fu">seq</span>(snap.files), <span class="cf">function</span>(i){</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">createSnap</span>(</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">file=</span>snap.files[i],</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">sample=</span>sample.names[i]</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>  );</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(x.sp.ls) <span class="ot">=</span> sample.names;</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>barcode.ls <span class="ot">=</span> <span class="fu">lapply</span>(<span class="fu">seq</span>(snap.files), <span class="cf">function</span>(i){</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>  barcodes <span class="ot">=</span> <span class="fu">read.csv</span>(</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>      barcode.files[i], </span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">head=</span><span class="cn">TRUE</span></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>  );</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>  barcodes <span class="ot">=</span> barcodes[<span class="dv">2</span><span class="sc">:</span><span class="fu">nrow</span>(barcodes),];</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a>  barcodes<span class="sc">$</span>logUMI <span class="ot">=</span> <span class="fu">log10</span>(barcodes<span class="sc">$</span>passed_filters <span class="sc">+</span> <span class="dv">1</span>);</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>  barcodes<span class="sc">$</span>promoter_ratio <span class="ot">=</span> (barcodes<span class="sc">$</span>promoter_region_fragments<span class="sc">+</span><span class="dv">1</span>) <span class="sc">/</span> (barcodes<span class="sc">$</span>passed_filters <span class="sc">+</span> <span class="dv">1</span>);</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>  barcodes</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>x.sp.ls</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for both datasets, we identify usable barcodes using [3.5-5] for log10(UMI) and [0.4-0.8] for promoter ratio as cutoff.</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>cutoff.logUMI.low <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">3.5</span>, <span class="fl">3.5</span>);</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>cutoff.logUMI.high <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">5</span>);</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>cutoff.FRIP.low <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.4</span>);</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>cutoff.FRIP.high <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">0.8</span>);</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>barcode.ls <span class="ot">=</span> <span class="fu">lapply</span>(<span class="fu">seq</span>(snap.files), <span class="cf">function</span>(i){</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  barcodes <span class="ot">=</span> barcode.ls[[i]];</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">=</span> <span class="fu">which</span>(</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>      barcodes<span class="sc">$</span>logUMI <span class="sc">&gt;=</span> cutoff.logUMI.low[i] <span class="sc">&amp;</span> </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>      barcodes<span class="sc">$</span>logUMI <span class="sc">&lt;=</span> cutoff.logUMI.high[i] <span class="sc">&amp;</span> </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>      barcodes<span class="sc">$</span>promoter_ratio <span class="sc">&gt;=</span> cutoff.FRIP.low[i] <span class="sc">&amp;</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>      barcodes<span class="sc">$</span>promoter_ratio <span class="sc">&lt;=</span> cutoff.FRIP.high[i]</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  );</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  barcodes[idx,]</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>});</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>x.sp.ls <span class="ot">=</span> <span class="fu">lapply</span>(<span class="fu">seq</span>(snap.files), <span class="cf">function</span>(i){</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  barcodes <span class="ot">=</span> barcode.ls[[i]];</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  x.sp <span class="ot">=</span> x.sp.ls[[i]];</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>  barcode.shared <span class="ot">=</span> <span class="fu">intersect</span>(x.sp<span class="sc">@</span>barcode, barcodes<span class="sc">$</span>barcode);</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>  x.sp <span class="ot">=</span> x.sp[<span class="fu">match</span>(barcode.shared, x.sp<span class="sc">@</span>barcode),];</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>  barcodes <span class="ot">=</span> barcodes[<span class="fu">match</span>(barcode.shared, barcodes<span class="sc">$</span>barcode),];</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>  x.sp<span class="sc">@</span>metaData <span class="ot">=</span> barcodes;</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>  x.sp</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(x.sp.ls) <span class="ot">=</span> sample.names;</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>x.sp.ls</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># combine two snap object</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> <span class="fu">Reduce</span>(snapRbind, x.sp.ls);</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>x.sp<span class="sc">@</span>metaData[<span class="st">&quot;Sample&quot;</span>] <span class="ot">=</span> x.sp<span class="sc">@</span>sample;</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">table</span>(x.sp<span class="sc">@</span>sample))</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>x.sp</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2. Add cell-by-bin matrix</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> <span class="fu">addBmatToSnap</span>(x.sp, <span class="at">bin.size=</span><span class="dv">5000</span>);</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 3. Matrix binarization</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> <span class="fu">makeBinary</span>(x.sp, <span class="at">mat=</span><span class="st">&quot;bmat&quot;</span>);</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 4. Bin filtering</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges);</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>black_list <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;hg19.blacklist.bed.gz&quot;</span>);</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>black_list.gr <span class="ot">=</span> <span class="fu">GRanges</span>(</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  black_list[,<span class="dv">1</span>], </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">IRanges</span>(black_list[,<span class="dv">2</span>], black_list[,<span class="dv">3</span>])</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>idy <span class="ot">=</span> <span class="fu">queryHits</span>(</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">findOverlaps</span>(x.sp<span class="sc">@</span>feature, black_list.gr)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(idy) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>  x.sp <span class="ot">=</span> x.sp[,<span class="sc">-</span>idy, mat<span class="ot">=</span><span class="st">&quot;bmat&quot;</span>];</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>};</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>x.sp</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove unwanted chromosomes</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>chr.exclude <span class="ot">=</span> <span class="fu">seqlevels</span>(x.sp<span class="sc">@</span>feature)[<span class="fu">grep</span>(<span class="st">&quot;random|chrM&quot;</span>, <span class="fu">seqlevels</span>(x.sp<span class="sc">@</span>feature))];</span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>idy <span class="ot">=</span> <span class="fu">grep</span>(<span class="fu">paste</span>(chr.exclude, <span class="at">collapse=</span><span class="st">&quot;|&quot;</span>), x.sp<span class="sc">@</span>feature);</span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(idy) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>  x.sp <span class="ot">=</span> x.sp[,<span class="sc">-</span>idy, mat<span class="ot">=</span><span class="st">&quot;bmat&quot;</span>]</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>};</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>x.sp</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The coverage of bins roughly obeys a log normal distribution. We remove the top 5% bins that overlap with invariant features such as the house keeping gene promoters.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>bin.cov <span class="ot">=</span> <span class="fu">log10</span>(Matrix<span class="sc">::</span><span class="fu">colSums</span>(x.sp<span class="sc">@</span>bmat)<span class="sc">+</span><span class="dv">1</span>);</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  bin.cov[bin.cov <span class="sc">&gt;</span> <span class="dv">0</span>], </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab=</span><span class="st">&quot;log10(bin cov)&quot;</span>, </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">main=</span><span class="st">&quot;log10(Bin Cov)&quot;</span>, </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">col=</span><span class="st">&quot;lightblue&quot;</span>, </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">5</span>)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>bin.cutoff <span class="ot">=</span> <span class="fu">quantile</span>(bin.cov[bin.cov <span class="sc">&gt;</span> <span class="dv">0</span>], <span class="fl">0.95</span>);</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>idy <span class="ot">=</span> <span class="fu">which</span>(bin.cov <span class="sc">&lt;=</span> bin.cutoff <span class="sc">&amp;</span> bin.cov <span class="sc">&gt;</span> <span class="dv">0</span>);</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> x.sp[, idy, mat<span class="ot">=</span><span class="st">&quot;bmat&quot;</span>];</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>x.sp</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="co"># We will further remove any cells of bin coverage less than 1,000. The rational behind this is that some cells may have high number of unique fragments but end up with low bin coverage after filtering. This step is optional but highly recommended.</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">=</span> <span class="fu">which</span>(Matrix<span class="sc">::</span><span class="fu">rowSums</span>(x.sp<span class="sc">@</span>bmat) <span class="sc">&gt;</span> <span class="dv">1000</span>);</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> x.sp[idx,];</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>x.sp</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 5. Dimensionality reduction</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>row.covs.dens <span class="ot">&lt;-</span> <span class="fu">density</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> x.sp<span class="sc">@</span>metaData[,<span class="st">&quot;logUMI&quot;</span>], </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">bw =</span> <span class="st">&#39;nrd&#39;</span>, <span class="at">adjust =</span> <span class="dv">1</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>sampling_prob <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="fu">approx</span>(<span class="at">x =</span> row.covs.dens<span class="sc">$</span>x, <span class="at">y =</span> row.covs.dens<span class="sc">$</span>y, <span class="at">xout =</span> x.sp<span class="sc">@</span>metaData[,<span class="st">&quot;logUMI&quot;</span>])<span class="sc">$</span>y <span class="sc">+</span> .Machine<span class="sc">$</span>double.eps);</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>);</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>idx.landmark.ds <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">sample</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="fu">nrow</span>(x.sp)), <span class="at">size =</span> <span class="dv">10000</span>, <span class="at">prob =</span> sampling_prob));</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>x.landmark.sp <span class="ot">=</span> x.sp[idx.landmark.ds,];</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>x.query.sp <span class="ot">=</span> x.sp[<span class="sc">-</span>idx.landmark.ds,];</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>x.landmark.sp <span class="ot">=</span> <span class="fu">runDiffusionMaps</span>(</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj=</span> x.landmark.sp,</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">input.mat=</span><span class="st">&quot;bmat&quot;</span>, </span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">num.eigs=</span><span class="dv">50</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>x.landmark.sp<span class="sc">@</span>metaData<span class="sc">$</span>landmark <span class="ot">=</span> <span class="dv">1</span>;</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>x.query.sp <span class="ot">=</span> <span class="fu">runDiffusionMapsExtension</span>(</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj1=</span>x.landmark.sp, </span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj2=</span>x.query.sp,</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">input.mat=</span><span class="st">&quot;bmat&quot;</span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>x.query.sp<span class="sc">@</span>metaData<span class="sc">$</span>landmark <span class="ot">=</span> <span class="dv">0</span>;</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> <span class="fu">snapRbind</span>(x.landmark.sp, x.query.sp);</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> x.sp[<span class="fu">order</span>(x.sp<span class="sc">@</span>metaData[<span class="st">&quot;sample&quot;</span>])];</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>x.sp <span class="ot">=</span> <span class="fu">runKNN</span>(</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj=</span>x.sp,</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">eigs.dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>,</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">k=</span><span class="dv">15</span></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>);</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>x.sp<span class="ot">=</span><span class="fu">runCluster</span>(</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj=</span>x.sp,</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">tmp.folder=</span><span class="fu">tempdir</span>(),</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">louvain.lib=</span><span class="st">&quot;R-igraph&quot;</span>, <span class="co">#&quot;leiden&quot; preferred, but may cause issues. Requires &#39;library(leiden)&#39;.</span></span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed.use=</span><span class="dv">10</span>,</span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">resolution=</span><span class="fl">0.7</span></span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="format-metadata" class="section level3">
<h3>Format Snap File Metadata</h3>
<p>Add the computed clusters to the Snap object metadata.</p>
<p>The Snap object contains two samples, “PBMC 5K” and “PBMC 10K”. Let’s
add a “Sample” column to the metadata. Let’s also add a column “files”
pointing to the original .snap files from which each cell came.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add clusters (from SnapATAC::runCluster) to metadata</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>x.sp<span class="sc">@</span>metaData<span class="sc">$</span>cluster <span class="ot">=</span> x.sp<span class="sc">@</span>cluster</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add Sample name to metadata (if not done previously)</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>x.sp<span class="sc">@</span>metaData<span class="sc">$</span>Sample <span class="ot">=</span> x.sp<span class="sc">@</span>sample</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Add files to metadata, indicating the original snap file each cell belongs to.</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>snap.files <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;atac_pbmc_5k_nextgem.snap&quot;</span>, </span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;atac_pbmc_10k_nextgem.snap&quot;</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>fileList <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(x.sp<span class="sc">@</span>metaData<span class="sc">$</span>Sample, <span class="cf">function</span>(x){</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ifelse</span>(x <span class="sc">==</span> <span class="st">&quot;PBMC 5K&quot;</span>, snap.files[[<span class="dv">1</span>]],snap.files[[<span class="dv">2</span>]])</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>x.sp<span class="sc">@</span>metaData<span class="sc">$</span>files <span class="ot">&lt;-</span> fileList</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="co"># SAVE this metadata to disk</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(x.sp<span class="sc">@</span>metaData, <span class="st">&quot;./snapMetadataforMOCHA.csv&quot;</span>)</span></code></pre></div>
</div>
<div id="extract-fragments" class="section level3">
<h3>Extract Fragments from Snap File</h3>
<p>Now we have a Snap object with metadata containing barcodes, unique
cell ids (column cell_id), sample names, and cell populations (cluster).
We also have our HG19 blackList, <code>black_list.gr</code>.</p>
<blockquote>
<p>Note: Following the tutorial from SnapATAC can often result in a
segfault when extracting fragments with
<code>SnapATAC::extractReads</code>. We recommend running on a machine
with large RAM and avoiding parallelization.</p>
</blockquote>
<p>Next we extract reads by sample and cell population, ensuring our
final GRanges list is named following the pattern
<code>CellPopulation#Sample</code>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>snapMetadata <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;./snapMetadataforMOCHA.csv&quot;</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>cellCol <span class="ot">&lt;-</span> <span class="st">&quot;barcode&quot;</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>cellPopLabel <span class="ot">&lt;-</span> <span class="st">&quot;cluster&quot;</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>cellPopulations <span class="ot">&lt;-</span> <span class="fu">unique</span>(snapMetadata<span class="sc">$</span>cluster)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>allSamples <span class="ot">&lt;-</span> <span class="fu">unique</span>(snapMetadata<span class="sc">$</span>Sample)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>fragmentsGRangesList <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(allSamples, <span class="cf">function</span>(sample){</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  barcodesList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(cellPopulations, <span class="cf">function</span>(cellPop) {</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    snapMetadata[snapMetadata<span class="sc">$</span>Sample <span class="sc">==</span> sample,]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Extract barcodes for a single cell population</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>    cellPopIdx <span class="ot">&lt;-</span> snapMetadata<span class="sc">$</span>cluster <span class="sc">==</span> cellPop</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>    cellPopBarcodes <span class="ot">&lt;-</span> snapMetadata[cellPopIdx,]<span class="sc">$</span>barcode</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Build the file list for the selected cell barcode</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>    files <span class="ot">&lt;-</span> snapMetadata[cellPopIdx,]<span class="sc">$</span>files</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Extract fragments</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>    cellPopFrags <span class="ot">&lt;-</span> SnapATAC<span class="sc">::</span><span class="fu">extractReads</span>(cellPopBarcodes, files, <span class="at">do.par =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(barcodesList) <span class="ot">&lt;-</span> <span class="fu">paste</span>(cellPopulations, sample, <span class="at">sep=</span><span class="st">&quot;#&quot;</span>)</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>  barcodesList</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>}))</span></code></pre></div>
<p>Calculate the study signal (the median number of fragments per
cell)</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>avg_reads <span class="ot">&lt;-</span> <span class="fu">lapply</span>(fragmentsGRangesList, <span class="cf">function</span>(x){</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  filtFrag <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="fu">as.data.frame</span>(x), barcode  <span class="sc">%in%</span> snapMetadata<span class="sc">$</span>barcode)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.vector</span>(<span class="fu">table</span>(filtFrag<span class="sc">$</span>barcode))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>studySignal <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">unlist</span>(avg_reads))</span></code></pre></div>
</div>
<div id="call-tiles" class="section level3">
<h3>Call Open Tiles with MOCHA</h3>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call Open Tiles</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>tileResults <span class="ot">&lt;-</span> MOCHA<span class="sc">::</span><span class="fu">callOpenTiles</span>(</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ATACFragments =</span> fragmentsGRangesList,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellColData =</span> snapMetadata, </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">blackList =</span> black_list.gr, </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">genome =</span> <span class="st">&quot;BSgenome.Hsapiens.UCSC.hg38&quot;</span>,</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellPopLabel =</span> cellPopLabel,</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellPopulations =</span> cellPopulations,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">studySignal =</span> studySignal,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellCol =</span> cellCol,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">TxDb =</span> <span class="st">&quot;TxDb.Hsapiens.UCSC.hg38.refGene&quot;</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">Org =</span> <span class="st">&quot;org.Hs.eg.db&quot;</span>, </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">outDir =</span> <span class="fu">paste</span>(<span class="fu">getwd</span>(),<span class="st">&#39;/MOCHA_Out&#39;</span>, <span class="at">sep =</span> <span class="st">&#39;&#39;</span>), </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">numCores =</span> <span class="dv">5</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>TSAM <span class="ot">&lt;-</span> MOCHA<span class="sc">::</span><span class="fu">getSampleTileMatrix</span>(tileResults, <span class="at">threshold =</span> <span class="fl">0.2</span>, <span class="at">numCores =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
