<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content='A statistical framework and analysis tool for open chromatin 
  analysis designed specifically for single cell ATAC-seq (Assay for 
  Transposase-Accessible Chromatin) data, after cell type/cluster 
  identification. These novel modules remove unwanted technical variation, 
  identify open chromatin, robustly models repeated measures in single cell 
  data, implement advanced statistical frameworks to model zero-inflation for 
  differential and co-accessibility analyses, and integrate with existing 
  databases and modules for downstream analyses to reveal biological insights. 
  MOCHA provides a statistical foundation for complex downstream analysis to 
  help advance the potential of single cell ATAC-seq for applied studies. 
  Methods for zero-inflated statistics are as described in: 
  Ghazanfar, S., Lin, Y., Su, X. et al. (2020) &lt;doi:10.1038/s41592-020-0885-x&gt;.
  Pimentel, Ronald Silva, "Kendalls Tau and Spearmans Rho for Zero-Inflated 
  Data" (2009) &lt;https://scholarworks.wmich.edu/dissertations/721/&gt;.'>
<title>Modeling for Single-Cell Open Chromatin Analysis â€¢ MOCHA</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Modeling for Single-Cell Open Chromatin Analysis">
<meta property="og:description" content='A statistical framework and analysis tool for open chromatin 
  analysis designed specifically for single cell ATAC-seq (Assay for 
  Transposase-Accessible Chromatin) data, after cell type/cluster 
  identification. These novel modules remove unwanted technical variation, 
  identify open chromatin, robustly models repeated measures in single cell 
  data, implement advanced statistical frameworks to model zero-inflation for 
  differential and co-accessibility analyses, and integrate with existing 
  databases and modules for downstream analyses to reveal biological insights. 
  MOCHA provides a statistical foundation for complex downstream analysis to 
  help advance the potential of single cell ATAC-seq for applied studies. 
  Methods for zero-inflated statistics are as described in: 
  Ghazanfar, S., Lin, Y., Su, X. et al. (2020) &lt;doi:10.1038/s41592-020-0885-x&gt;.
  Pimentel, Ronald Silva, "Kendalls Tau and Spearmans Rho for Zero-Inflated 
  Data" (2009) &lt;https://scholarworks.wmich.edu/dissertations/721/&gt;.'>
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">MOCHA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/COVID-walkthrough.html">Walkthrough with COVID Dataset</a>
    <a class="dropdown-item" href="articles/ImportingFromOtherSources.html">Importing From Other Sources</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aifimmunology/MOCHA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="mocha-model-based-single-cell-open-chromatin-analysis">MOCHA: Model-based single-cell Open Chromatin Analysis<a class="anchor" aria-label="anchor" href="#mocha-model-based-single-cell-open-chromatin-analysis"></a>
</h1></div>
<p>MOCHA is an R package meant to enable robust sample-specific scATAC analysis by providing a novel single-cell peak-calling algorithm and leveraging the latest in zero-inflated statistical methods. We hope this package will enable translational research on scATAC datasets by addressing statistical issues around peak calling, differential accessibility, co-accessibility, and repeated measures. Our goal is to enable intuitive and robust identification of open chromatin, differential accessibility, and co-accessibility so that biomedical researchers can leverage scATAC datasets to identify the therapeutic targets.</p>
<p>Find out more by visiting the <a href="https://aifimmunology.github.io/MOCHA/">MOCHA website</a>.</p>
<hr>
<div class="section level3">
<h3 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h3>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#contact">Contact</a></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="id_-installation">
<a name="installation"></a> Installation<a class="anchor" aria-label="anchor" href="#id_-installation"></a>
</h2>
<p>Install from binaries (stable release on CRAN):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"MOCHA"</span>)</span></code></pre></div>
<p>Install from source:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"aifimmunology/MOCHA"</span>)</span></code></pre></div>
<p>Install a specific development branch from source:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"aifimmunology/MOCHA"</span>, <span class="at">ref =</span> <span class="st">"your_branch_name"</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="id_-usage-overview">
<a name="overview"></a> Usage Overview<a class="anchor" aria-label="anchor" href="#id_-usage-overview"></a>
</h2>
<p>Please view the example usage found in the vignette found under <code>vignettes/COVID-walkthrough.html</code>.</p>
<p>The example usage demonstrates this workflow:</p>
<div class="figure">
<img src="reference/figures/workflow_diagram_v3_website.png" alt=""><p class="caption">Workflow</p>
</div>
</div>
<div class="section level2">
<h2 id="id_-contact">
<a name="contact"></a> Contact<a class="anchor" aria-label="anchor" href="#id_-contact"></a>
</h2>
<p>While the pipeline can be run function-to-function, you may wish to inspect intermediate results or use end results for your own custom analyses. All MOCHA outputs use common Bioconductor data structures. We also provide some getters for accessing specific results.</p>
</div>
<div class="section level2">
<h2 id="callopentiles-results">callOpenTiles results<a class="anchor" aria-label="anchor" href="#callopentiles-results"></a>
</h2>
<p>The output of the first step of the pipeline, <code><a href="reference/callOpenTiles-methods.html">MOCHA::callOpenTiles</a></code> is a <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html#overview-of-the-multiassayexperiment-class" class="external-link">MultiAssayExperiment</a> organizing your open-tiles by cell population. This is the input to the next function, <code>MOCHA::getSampleTileMatrices</code>.</p>
<p>Sample-level metadata can be accessed using <code>colData(tileResults)</code>.</p>
<p>A specific cell populationâ€™s results are stored as a <a href="https://bioconductor.org/packages/release/bioc/vignettes/RaggedExperiment/inst/doc/RaggedExperiment.html" class="external-link">RaggedExperiment</a>.</p>
<p>This object stores ranged data alongside the sample-level metadata.</p>
</div>
<div class="section level2">
<h2 id="getsampletilematrix-results">getSampleTileMatrix results<a class="anchor" aria-label="anchor" href="#getsampletilematrix-results"></a>
</h2>
<p>The output of getSampleTileMatrix is a <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a> organizing the TSAM (Tile Sample Accessibility Matrix) by cell population. This is the input to <code><a href="reference/getDifferentialAccessibleTiles.html">MOCHA::getDifferentialAccessibleTiles</a></code> and other downstream analyses.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Subset results for Chr 4</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> SampleTileMatrices[<span class="fu">grep</span>(<span class="st">'chr4'</span>, <span class="fu">rownames</span>(SampleTileMatrices)),]</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>class<span class="sc">:</span> RangedSummarizedExperiment </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>dim<span class="sc">:</span> <span class="dv">8112</span> <span class="dv">39</span> </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(<span class="dv">6</span>)<span class="sc">:</span> CellCounts FragmentCounts ... OrgDb Directory</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(<span class="dv">1</span>)<span class="sc">:</span> CD16 Mono</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(<span class="dv">8112</span>)<span class="sc">:</span> chr4<span class="sc">:</span><span class="dv">10005500-10005999</span> chr4<span class="sc">:</span><span class="dv">10006000-10006499</span> ...</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  chr4<span class="sc">:</span><span class="dv">99961000-99961499</span> chr4<span class="sc">:</span><span class="dv">99961500-99961999</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>rowData <span class="fu">names</span>(<span class="dv">1</span>)<span class="sc">:</span> CD16 Mono</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="dv">39</span>)<span class="sc">:</span> B011<span class="sc">-</span>AP0C1W3 B011<span class="sc">-</span>AP0C1W8 ... FSQEAZ0C2D3<span class="dv">-02</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  FSQFAZ0BZJQ<span class="dv">-02</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>colData <span class="fu">names</span>(<span class="dv">186</span>)<span class="sc">:</span> Sample well_id ... ATAC_WellID AIFI.Batch</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span></code></pre></div>
<p>It also holds metadata related to the genome, transcript database, and annotations: <code>metadata(SampleTileMatrices)</code></p>
</div>
<div class="section level2">
<h2 id="getdifferentialaccessibletiles-results">getDifferentialAccessibleTiles results<a class="anchor" aria-label="anchor" href="#getdifferentialaccessibletiles-results"></a>
</h2>
<p>Results of <code><a href="reference/getDifferentialAccessibleTiles.html">MOCHA::getDifferentialAccessibleTiles</a></code> is given either as a <code>data.table</code> or a â€˜grangesâ€™ and can be filtered accordingly:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(plyranges<span class="sc">::</span><span class="fu">filter</span>(differentials, seqnames <span class="sc">==</span><span class="st">'chr4'</span> <span class="sc">&amp;</span> FDR <span class="sc">&lt;</span> <span class="fl">0.2</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>GRanges object with <span class="dv">6</span> ranges and <span class="dv">13</span> metadata columns<span class="sc">:</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>      seqnames              ranges strand <span class="sc">|</span>                   Tile</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>         <span class="sc">&lt;</span>Rle<span class="sc">&gt;</span>           <span class="er">&lt;</span>IRanges<span class="sc">&gt;</span>  <span class="er">&lt;</span>Rle<span class="sc">&gt;</span> <span class="er">|</span>            <span class="er">&lt;</span>character<span class="sc">&gt;</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">1</span>]     chr4   <span class="dv">10023500-10023999</span>      <span class="sc">*</span> <span class="er">|</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">2</span>]     chr4   <span class="dv">10024000-10024499</span>      <span class="sc">*</span> <span class="er">|</span> chr4<span class="sc">:</span><span class="dv">10024000-10024499</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">3</span>]     chr4   <span class="dv">10063500-10063999</span>      <span class="sc">*</span> <span class="er">|</span> chr4<span class="sc">:</span><span class="dv">10063500-10063999</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">4</span>]     chr4   <span class="dv">10096500-10096999</span>      <span class="sc">*</span> <span class="er">|</span> chr4<span class="sc">:</span><span class="dv">10096500-10096999</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">5</span>]     chr4     <span class="dv">1010500-1010999</span>      <span class="sc">*</span> <span class="er">|</span>   chr4<span class="sc">:</span><span class="dv">1010500-1010999</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">6</span>]     chr4 <span class="dv">101346000-101346499</span>      <span class="sc">*</span> <span class="er">|</span> chr4<span class="sc">:</span><span class="dv">101346000-10134</span>..</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>      CellPopulation  Foreground  Background     P_value Test_Statistic</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>         <span class="sc">&lt;</span>character<span class="sc">&gt;</span> <span class="er">&lt;</span>character<span class="sc">&gt;</span> <span class="er">&lt;</span>character<span class="sc">&gt;</span>   <span class="er">&lt;</span>numeric<span class="sc">&gt;</span>      <span class="er">&lt;</span>numeric<span class="sc">&gt;</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">1</span>]      CD16 Mono    Positive    Negative <span class="fl">0.002610899</span>       <span class="fl">11.89612</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">2</span>]      CD16 Mono    Positive    Negative <span class="fl">0.023560228</span>        <span class="fl">7.49639</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">3</span>]      CD16 Mono    Positive    Negative <span class="fl">0.000421733</span>       <span class="fl">12.43336</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">4</span>]      CD16 Mono    Positive    Negative <span class="fl">0.006561717</span>       <span class="fl">10.05301</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">5</span>]      CD16 Mono    Positive    Negative <span class="fl">0.016701496</span>        <span class="fl">5.72747</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">6</span>]      CD16 Mono    Positive    Negative <span class="fl">0.006274534</span>        <span class="fl">7.46972</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>            FDR  Log2FC_C   MeanDiff Avg_Intensity_Case Pct0_Case</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>      <span class="sc">&lt;</span>numeric<span class="sc">&gt;</span> <span class="er">&lt;</span>numeric<span class="sc">&gt;</span>  <span class="er">&lt;</span>numeric<span class="sc">&gt;</span>          <span class="er">&lt;</span>numeric<span class="sc">&gt;</span> <span class="er">&lt;</span>numeric<span class="sc">&gt;</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">1</span>] <span class="fl">0.0788621</span>  <span class="fl">1.023296</span>  <span class="fl">2.9715709</span>            <span class="fl">12.3478</span> <span class="fl">0.0000000</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">2</span>] <span class="fl">0.1910923</span>  <span class="fl">0.648633</span> <span class="sc">-</span><span class="fl">0.0879542</span>            <span class="fl">12.2027</span> <span class="fl">0.0588235</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">3</span>] <span class="fl">0.0451423</span>  <span class="fl">0.786501</span>  <span class="fl">0.7773734</span>            <span class="fl">12.8422</span> <span class="fl">0.0000000</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">4</span>] <span class="fl">0.1124369</span>  <span class="fl">0.739042</span>  <span class="fl">1.3575321</span>            <span class="fl">13.0183</span> <span class="fl">0.0000000</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">5</span>] <span class="fl">0.1628641</span> <span class="sc">-</span><span class="fl">0.476150</span> <span class="sc">-</span><span class="fl">0.4293368</span>            <span class="fl">13.6077</span> <span class="fl">0.0000000</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">6</span>] <span class="fl">0.1091473</span> <span class="sc">-</span><span class="fl">0.397274</span> <span class="sc">-</span><span class="fl">0.3913687</span>            <span class="fl">13.1689</span> <span class="fl">0.0000000</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>      Avg_Intensity_Control Pct0_Control</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>                  <span class="sc">&lt;</span>numeric<span class="sc">&gt;</span>    <span class="er">&lt;</span>numeric<span class="sc">&gt;</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">1</span>]               <span class="fl">11.1988</span>    <span class="fl">0.1818182</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">2</span>]               <span class="fl">11.4155</span>    <span class="fl">0.0000000</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">3</span>]               <span class="fl">12.0881</span>    <span class="fl">0.0000000</span></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">4</span>]               <span class="fl">12.2886</span>    <span class="fl">0.0454545</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">5</span>]               <span class="fl">14.2039</span>    <span class="fl">0.0000000</span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">6</span>]               <span class="fl">13.5651</span>    <span class="fl">0.0000000</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>  <span class="sc">-------</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>  seqinfo<span class="sc">:</span> <span class="dv">24</span> sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getcoaccessiblelinks-results">getCoAccessibleLinks results<a class="anchor" aria-label="anchor" href="#getcoaccessiblelinks-results"></a>
</h2>
<p>Results of <code><a href="reference/getCoAccessibleLinks.html">MOCHA::getCoAccessibleLinks</a></code> is given as a <code>data.frame</code> and can be filtered further according to correlation using <code><a href="reference/filterCoAccessibleLinks.html">MOCHA::filterCoAccessibleLinks</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(links)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>   Correlation                  Tile1                  Tile2</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span>  <span class="fl">0.06203363</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10005500-10005999</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span><span class="sc">:</span> <span class="sc">-</span><span class="fl">0.07812306</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10006000-10006499</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">:</span>  <span class="fl">0.06192133</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10006500-10006999</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span><span class="sc">:</span>  <span class="fl">0.25329153</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10007000-10007499</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span><span class="sc">:</span>  <span class="fl">0.09938466</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10013500-10013999</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span><span class="sc">:</span>  <span class="fl">0.03643015</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10015000-10015499</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> MOCHA<span class="sc">::</span><span class="fu">filterCoAccessibleLinks</span>(links, <span class="at">threshold =</span> <span class="fl">0.4</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>   Correlation                  Tile1                  Tile2  chr    start      end</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span>   <span class="fl">0.5035455</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10024000-10024499</span> chr4 <span class="dv">10023500</span> <span class="dv">10024499</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span><span class="sc">:</span>   <span class="fl">0.5121115</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10063500-10063999</span> chr4 <span class="dv">10023500</span> <span class="dv">10063999</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">:</span>   <span class="fl">0.4059254</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10078000-10078499</span> chr4 <span class="dv">10023500</span> <span class="dv">10078499</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span><span class="sc">:</span>   <span class="fl">0.4003500</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10096500-10096999</span> chr4 <span class="dv">10023500</span> <span class="dv">10096999</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span><span class="sc">:</span>   <span class="fl">0.4367843</span> chr4<span class="sc">:</span><span class="dv">10023500-10023999</span> chr4<span class="sc">:</span><span class="dv">10106500-10106999</span> chr4 <span class="dv">10023500</span> <span class="dv">10106999</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="id_-contact-1">
<a name="contact"></a> Contact<a class="anchor" aria-label="anchor" href="#id_-contact-1"></a>
</h1>
<p>To contact the developers on issues and feature requests, please contact us via the discussions tab for feature requests, or open issues for any bugs.</p>
</div>
<div class="section level1">
<h1 id="id_-license">
<a name="license"></a> License<a class="anchor" aria-label="anchor" href="#id_-license"></a>
</h1>
<p>MOCHA follows the Allen Institute Software License - full information about the license can be found on the LICENSE file.</p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=MOCHA" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/aifimmunology/MOCHA/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/aifimmunology/MOCHA/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MOCHA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Samir Rachid Zaim <br><small class="roles"> Author, contributor </small>  </li>
<li>Mark-Phillip Pebworth <br><small class="roles"> Author, contributor </small>  </li>
<li>Imran McGrath <br><small class="roles"> Author, maintainer </small>  </li>
<li>Lauren Okada <br><small class="roles"> Author, contributor </small>  </li>
<li>Xiaojun Li <br><small class="roles"> Author, contributor </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Samir Rachid Zaim, Mark-Phillip Pebworth, Imran McGrath, Lauren Okada, Xiaojun Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
