<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MOCHA">
<title>COVID-walkthrough • MOCHA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="COVID-walkthrough">
<meta property="og:description" content="MOCHA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MOCHA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/COVID-walkthrough.html">COVID-walkthrough</a>
    <a class="dropdown-item" href="../articles/Comparing-Cell-Populations.html">Comparing-Cell-Populations</a>
    <a class="dropdown-item" href="../articles/Dev-Lessons.html">Lessons from Development</a>
    <a class="dropdown-item" href="../articles/Results-Tips.html">Results-Tips</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aifimmunology/MOCHA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>COVID-walkthrough</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/aifimmunology/MOCHA/blob/HEAD/vignettes/COVID-walkthrough.Rmd" class="external-link"><code>vignettes/COVID-walkthrough.Rmd</code></a></small>
      <div class="d-none name"><code>COVID-walkthrough.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://aifimmunology.github.io/MOCHA/">MOCHA</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ArchR</span><span class="op">)</span></span></code></pre></div>
<div id="section" class="section level53">
<p class="heading"></p>
</div>
<div class="section level2">
<h2 id="load-libraries-archr-project-and-annotation-databases">0. Load libraries, ArchR project, and annotation databases<a class="anchor" aria-label="anchor" href="#load-libraries-archr-project-and-annotation-databases"></a>
</h2>
<p>Optionally filter the ArchR project to a subset of samples.</p>
<div id="section-1" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You should substitute this with your own ArchR project.</span></span>
<span><span class="co"># You must have completed cell labeling with your ArchR project.</span></span>
<span></span>
<span><span class="va">myArchRProj</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/loadArchRProject.html" class="external-link">loadArchRProject</a></span><span class="op">(</span><span class="st">"/home/jupyter/FullCovid"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define your annotation package for TxDb object(s)</span></span>
<span><span class="co"># and genome-wide annotation.</span></span>
<span><span class="co"># Here our samples are human using hg38 as a reference.</span></span>
<span><span class="co"># For more info: &lt;https://bioconductor.org/packages/3.15/data/annotation/&gt;</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.refGene</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span> </span>
<span><span class="va">TxDb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.refGene</span> </span>
<span><span class="va">Org</span> <span class="op">&lt;-</span> <span class="va">org.Hs.eg.db</span></span>
<span></span>
<span><span class="co"># Optional: Filter your ArchR project by sample.</span></span>
<span><span class="co"># For our example we filter ArchR Project to three samples from</span></span>
<span><span class="co"># each COVID Status (3 Positive, 3 Negative).</span></span>
<span></span>
<span><span class="va">samplesToKeep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"B011-AP0C1W3"</span>, <span class="st">"B011-AP0C1W8"</span>, <span class="st">"B011-AP0C2W1"</span>, <span class="st">"B025_FSQAAZ0BZZS-01"</span>, <span class="st">"B025_FSQAAZ0C0YJ-01"</span>, <span class="st">"B025_FSQAAZ0C00P-07"</span> <span class="op">)</span></span>
<span><span class="va">idxSample</span> <span class="op">&lt;-</span> <span class="fu">BiocGenerics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">myArchRProj</span><span class="op">$</span><span class="va">Sample</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">samplesToKeep</span><span class="op">)</span> </span>
<span><span class="va">cellsSample</span> <span class="op">&lt;-</span> <span class="va">myArchRProj</span><span class="op">$</span><span class="va">cellNames</span><span class="op">[</span><span class="va">idxSample</span><span class="op">]</span></span>
<span><span class="va">myArchRProj</span> <span class="op">&lt;-</span> <span class="va">myArchRProj</span><span class="op">[</span><span class="va">cellsSample</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div id="section-2" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="setting-parameters">1. Setting Parameters<a class="anchor" aria-label="anchor" href="#setting-parameters"></a>
</h2>
<p>These should be set according to your ArchR project and investigative
question.</p>
<p>For more details on each of these parameters, view the help pages for
each function using <code><a href="../reference/callOpenTiles-methods.html">?callOpenTiles</a></code> and
<code><a href="../reference/getSampleTileMatrix.html">?getSampleTileMatrix</a></code></p>
<div id="section-3" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Parameters for calling open tiles.</span></span>
<span><span class="va">cellPopLabel</span> <span class="op">&lt;-</span> <span class="st">"CellSubsets"</span> </span>
<span><span class="va">cellPopulations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MAIT"</span>, <span class="st">"CD16 Mono"</span>, <span class="st">"DC"</span><span class="op">)</span> </span>
<span><span class="va">numCores</span> <span class="op">&lt;-</span> <span class="fl">20</span></span></code></pre></div>
</div>
<div id="section-4" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="call-open-tiles">2. Call open tiles<a class="anchor" aria-label="anchor" href="#call-open-tiles"></a>
</h2>
<p>Get sample-tile matrices for all specified cell populations.</p>
<div id="section-5" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tileResults</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/callOpenTiles-methods.html">callOpenTiles</a></span><span class="op">(</span></span>
<span>  <span class="va">myArchRProj</span>, </span>
<span>  cellPopLabel <span class="op">=</span> <span class="va">cellPopLabel</span>, </span>
<span>  cellPopulations <span class="op">=</span> <span class="va">cellPopulations</span>, </span>
<span>  TxDb <span class="op">=</span> <span class="va">TxDb</span>, </span>
<span>  Org <span class="op">=</span> <span class="va">Org</span>, </span>
<span>  numCores <span class="op">=</span> <span class="va">numCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-6" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="get-consensus-sample-tile-matrices">3. Get consensus sample-tile matrices<a class="anchor" aria-label="anchor" href="#get-consensus-sample-tile-matrices"></a>
</h2>
<p>…for all cell populations.</p>
<p>These matrices are organized by cell population
RangedSummarizedExperiment object and are the primary input to
downstream analyses. An advantage of MOCHA’s ability to call
sample-specific open tiles is that we can determine a high-quality set
of”consensus tiles determined as follows: each sample “votes” on whether
a tile is open for that sample, and we keep tiles that are called open
by a minimum percentage of samples. The minimum percentage of samples
which a tile must be called in to be retained is controlled by
<code>threshold</code>. <code>groupColumn</code> can be provided to
specify a metadata column that separates your data by sample groups,
e.g. if you have a case and control groups. Consensus tiles will be
computed for each group, and the union of consensus tiles from each
group are retained. This should be used where there are expected
biological differences between the sample groups. Currently it is best
utilized when each group has a similar size, as <code>threshold</code>
will be applied to evenly each group. By default, groupColumn is null
and all samples will be pooled to vote on consensus tiles.</p>
<div id="section-7" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We have 6 samples total: 3 samples for each COVID status (3 positive and 3 negative).</span></span>
<span><span class="co"># Since these groupings may have unique biology and we expect differences</span></span>
<span><span class="co"># in accessibility, we want to compute consensus tiles on each </span></span>
<span><span class="co"># group independently and take the union of consensus tiles from each group.</span></span>
<span><span class="va">groupColumn</span> <span class="op">&lt;-</span> <span class="st">"COVID_status"</span> </span>
<span></span>
<span><span class="co"># We set the threshold to require a tile must be open in at least 2 samples</span></span>
<span><span class="co"># in its group to be retained (2/3=0.66)</span></span>
<span><span class="va">threshold</span> <span class="op">&lt;-</span> <span class="fl">0.66</span></span>
<span></span>
<span><span class="co"># Alternatively, you can set the threshold to 0 to keep the union of</span></span>
<span><span class="co"># all samples' open tiles.</span></span>
<span><span class="co"># This is equivalent to setting a threshold that would retain</span></span>
<span><span class="co"># tiles that are open in at least one sample. </span></span>
<span></span>
<span><span class="va">SampleTileMatrices</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/getSampleTileMatrix.html">getSampleTileMatrix</a></span><span class="op">(</span></span>
<span>  <span class="va">tileResults</span>, </span>
<span>  cellPopulations <span class="op">=</span> <span class="va">cellPopulations</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>  groupColumn <span class="op">=</span> <span class="va">groupColumn</span>, </span>
<span>  threshold <span class="op">=</span> <span class="va">threshold</span>, </span>
<span>  log2Intensity <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-8" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="optional-add-gene-annotations">4. (Optional) Add gene annotations<a class="anchor" aria-label="anchor" href="#optional-add-gene-annotations"></a>
</h2>
<p>…and motifs to our SampleTileMatrices.</p>
<p>This info will aid further downstream analyses but is not required
for differential accessibility nor coaccessibility.</p>
<div id="section-9" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This function can also take any GRanges object</span></span>
<span><span class="co"># and add annotations to its metadata.</span></span>
<span><span class="va">SampleTileMatricesAnnotated</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/annotateTiles.html">annotateTiles</a></span><span class="op">(</span><span class="va">SampleTileMatrices</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load a curated motif set from library(chromVARmotifs)</span></span>
<span><span class="co"># included with ArchR installation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">human_pwms_v2</span><span class="op">)</span></span>
<span><span class="va">SampleTileMatricesAnnotated</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/addMotifSet.html">addMotifSet</a></span><span class="op">(</span></span>
<span>  <span class="va">SampleTileMatricesAnnotated</span>, </span>
<span>  pwms <span class="op">=</span> <span class="va">human_pwms_v2</span>, </span>
<span>  w <span class="op">=</span> <span class="fl">7</span> <span class="co"># weight parameter for motifmatchr</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-10" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="optional-plot-a-specific-regions-coverage">5. (Optional) Plot a specific region’s coverage<a class="anchor" aria-label="anchor" href="#optional-plot-a-specific-regions-coverage"></a>
</h2>
<p>Here we plot coverage at a specific region and gene by infection
stage.</p>
<div id="section-11" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">countSE</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/extractRegion.html">extractRegion</a></span><span class="op">(</span></span>
<span>  SampleTileObj <span class="op">=</span> <span class="va">SampleTileMatrices</span>, </span>
<span>  cellPopulations <span class="op">=</span> <span class="st">"CD16 Mono"</span>, </span>
<span>  region <span class="op">=</span> <span class="st">"chr3:38137866-38139912"</span>, </span>
<span>  groupColumn <span class="op">=</span> <span class="st">"COVID_status"</span>, </span>
<span>  numCores <span class="op">=</span> <span class="va">numCores</span>, </span>
<span>  sampleSpecific <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"ExamplePlot.pdf"</span><span class="op">)</span></span>
<span><span class="co"># Note that to show specific genes with the option' whichGene'</span></span>
<span><span class="co"># you must have the package RMariaDB installed </span></span>
<span><span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span>countSE <span class="op">=</span> <span class="va">countSE</span>, whichGene <span class="op">=</span> <span class="st">"MYD88"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-12" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="get-differential-accessibility-for-specific-cell-populations">6. Get differential accessibility for specific cell populations<a class="anchor" aria-label="anchor" href="#get-differential-accessibility-for-specific-cell-populations"></a>
</h2>
<p>Here we are comparing MAIT cells between samples where our
groupColumn “COVID_status” is Positive (our foreground) to Negative
samples (our background).</p>
<div id="section-13" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellPopulation</span> <span class="op">&lt;-</span> <span class="st">"MAIT"</span></span>
<span><span class="va">foreground</span> <span class="op">&lt;-</span> <span class="st">"Positive"</span></span>
<span><span class="va">background</span> <span class="op">&lt;-</span> <span class="st">"Negative"</span> </span>
<span></span>
<span><span class="co"># Standard output will display the number of tiles found below a false-discovery rate threshold.</span></span>
<span><span class="co"># This parameter does not filter results and only affects the afforementioned message. </span></span>
<span><span class="va">fdrToDisplay</span> <span class="op">&lt;-</span> <span class="fl">0.2</span></span>
<span></span>
<span><span class="co"># Choose to output a GRanges or data.frame.</span></span>
<span><span class="co"># Default is TRUE</span></span>
<span><span class="va">outputGRanges</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span></span>
<span><span class="va">differentials</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/getDifferentialAccessibleTiles.html">getDifferentialAccessibleTiles</a></span><span class="op">(</span></span>
<span>  SampleTileObj <span class="op">=</span> <span class="va">SampleTileMatrices</span>, </span>
<span>  cellPopulation <span class="op">=</span> <span class="va">cellPopulation</span>, </span>
<span>  groupColumn <span class="op">=</span> <span class="va">groupColumn</span>, </span>
<span>  foreground <span class="op">=</span> <span class="va">foreground</span>, </span>
<span>  background <span class="op">=</span> <span class="va">background</span>, </span>
<span>  fdrToDisplay <span class="op">=</span> <span class="va">fdrToDisplay</span>, </span>
<span>  outputGRanges <span class="op">=</span> <span class="va">outputGRanges</span>, </span>
<span>  numCores <span class="op">=</span> <span class="va">numCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-14" class="section level53">
<p class="heading"></p>
</div>
</div>
<div class="section level2">
<h2 id="get-co-accessible-links">7. Get co-accessible links<a class="anchor" aria-label="anchor" href="#get-co-accessible-links"></a>
</h2>
<p>…between input regions (tiles) and their neighboring regions within a
window.</p>
<p>Here we give the first ten differential tiles as our input
regions.</p>
<div id="section-15" class="section level53">
<p class="heading"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">differentials</span>, <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively, define regions as a character vector</span></span>
<span><span class="co"># of region strings in the format "chr:start-end"</span></span>
<span><span class="co"># regions &lt;- c(</span></span>
<span><span class="co"># "chrY:7326500-7326999",</span></span>
<span><span class="co"># "chrY:7327000-7327499",</span></span>
<span><span class="co"># "chrY:7339500-7339999",</span></span>
<span><span class="co"># "chrY:7344500-7344999"</span></span>
<span><span class="co"># )</span></span>
<span></span>
<span><span class="va">links</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/getCoAccessibleLinks.html">getCoAccessibleLinks</a></span><span class="op">(</span></span>
<span>  SampleTileObj <span class="op">=</span> <span class="va">SampleTileMatrices</span>,</span>
<span>  cellPopulation <span class="op">=</span> <span class="va">cellPopulation</span>,</span>
<span>  regions <span class="op">=</span> <span class="va">regions</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">1</span> <span class="op">*</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span>,</span>
<span>  numCores <span class="op">=</span> <span class="va">numCores</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Optionally filter these links by their absolute</span></span>
<span><span class="co"># correlation - this output also adds the chromosome,</span></span>
<span><span class="co"># start, and end site of each link to the table.</span></span>
<span></span>
<span><span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/filterCoAccessibleLinks.html">filterCoAccessibleLinks</a></span><span class="op">(</span><span class="va">links</span>, threshold <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Samir Rachid Zaim, Mark-Phillip Pebworth, Imran McGrath, Lauren Okada, Xiaojun Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
