<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MOCHA">
<title>Importing From Other Sources • MOCHA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Importing From Other Sources">
<meta property="og:description" content="MOCHA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MOCHA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/COVID-walkthrough.html">Walkthrough with COVID Dataset</a>
    <a class="dropdown-item" href="../articles/ImportingFromOtherSources.html">Importing From Other Sources</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aifimmunology/MOCHA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Importing From Other Sources</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/aifimmunology/MOCHA/blob/HEAD/vignettes/ImportingFromOtherSources.Rmd" class="external-link"><code>vignettes/ImportingFromOtherSources.Rmd</code></a></small>
      <div class="d-none name"><code>ImportingFromOtherSources.Rmd</code></div>
    </div>

    
    
<p>MOCHA is intended for use after clustering and cell labeling, since
all functions of MOCHA operate on per-sample basis within each cell
population. MOCHA can be used with an ArchR Project as input, as well as
with a generic input. This tutorial has two sections: demonstrating
MOCHA with generic inputs - <a href="#signac">Importing from Signac</a>
and <a href="#snapatac">Importing from SnapATAC</a>.</p>
<div class="section level2">
<h2 id="signac">Importing from Signac<a class="anchor" aria-label="anchor" href="#signac"></a>
</h2>
<p>This vignette will follow the <a href="https://stuartlab.org/signac/articles/pbmc_multiomic.html" class="external-link">Signac
tutorial</a> to generate an Signac object and label cell types within
it. If you already have a Signac object with cell types labelled, skip
to section <a href="#extract-frags-signac">Extract Fragments from
Signac</a></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratDisk</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="signac-tutorial-through-clustering">Signac Tutorial Through Clustering<a class="anchor" aria-label="anchor" href="#signac-tutorial-through-clustering"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load in ATAC and RNA data</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">fragpath</span> <span class="op">&lt;-</span> <span class="st">"pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a Seurat object containing the RNA adata</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span><span class="op">$</span><span class="va">`Gene Expression`</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create ATAC assay and add it to the object</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CreateChromatinAssay.html" class="external-link">CreateChromatinAssay</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span><span class="op">$</span><span class="va">Peaks</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>  fragments <span class="op">=</span> <span class="va">fragpath</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">annotation</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/NucleosomeSignal.html" class="external-link">NucleosomeSignal</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/TSSEnrichment.html" class="external-link">TSSEnrichment</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  subset <span class="op">=</span> <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="fl">100000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">25000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_RNA</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nucleosome_signal</span> <span class="op">&lt;</span> <span class="fl">2</span> <span class="op">&amp;</span></span>
<span>    <span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transform RNA </span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span></code></pre></div>
<p>Label cell types:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load PBMC reference</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'wget https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat'</span><span class="op">)</span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu">LoadH5Seurat</span><span class="op">(</span><span class="st">"pbmc_multimodal.h5seurat"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span>
<span></span>
<span><span class="co"># transfer cell type labels from reference to query</span></span>
<span><span class="va">transfer_anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindTransferAnchors.html" class="external-link">FindTransferAnchors</a></span><span class="op">(</span></span>
<span>  reference <span class="op">=</span> <span class="va">reference</span>,</span>
<span>  query <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  normalization.method <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  reference.reduction <span class="op">=</span> <span class="st">"spca"</span>,</span>
<span>  recompute.residuals <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/TransferData.html" class="external-link">TransferData</a></span><span class="op">(</span></span>
<span>  anchorset <span class="op">=</span> <span class="va">transfer_anchors</span>, </span>
<span>  refdata <span class="op">=</span> <span class="va">reference</span><span class="op">$</span><span class="va">celltype.l2</span>,</span>
<span>  weight.reduction <span class="op">=</span> <span class="va">pbmc</span><span class="op">[[</span><span class="st">'pca'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AddMetaData.html" class="external-link">AddMetaData</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">predictions</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set the cell identities to the cell type predictions</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"predicted.id"</span></span>
<span></span>
<span><span class="co"># set a reasonable order for cell types to be displayed when plotting</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4 Naive"</span>, <span class="st">"CD4 TCM"</span>, <span class="st">"CD4 CTL"</span>, <span class="st">"CD4 TEM"</span>, <span class="st">"CD4 Proliferating"</span>,</span>
<span>                  <span class="st">"CD8 Naive"</span>, <span class="st">"dnT"</span>,</span>
<span>                 <span class="st">"CD8 TEM"</span>, <span class="st">"CD8 TCM"</span>, <span class="st">"CD8 Proliferating"</span>, <span class="st">"MAIT"</span>, <span class="st">"NK"</span>, <span class="st">"NK_CD56bright"</span>,</span>
<span>                 <span class="st">"NK Proliferating"</span>, <span class="st">"gdT"</span>,</span>
<span>                 <span class="st">"Treg"</span>, <span class="st">"B naive"</span>, <span class="st">"B intermediate"</span>, <span class="st">"B memory"</span>, <span class="st">"Plasmablast"</span>,</span>
<span>                 <span class="st">"CD14 Mono"</span>, <span class="st">"CD16 Mono"</span>,</span>
<span>                 <span class="st">"cDC1"</span>, <span class="st">"cDC2"</span>, <span class="st">"pDC"</span>, <span class="st">"HSPC"</span>, <span class="st">"Eryth"</span>, <span class="st">"ASDC"</span>, <span class="st">"ILC"</span>, <span class="st">"Platelet"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">'pbmc_Signac_tutorial.rds'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-frags-signac">Extract Fragments from Signac Object<a class="anchor" aria-label="anchor" href="#extract-frags-signac"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'pbmc_Signac_tutorial.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'ATAC'</span></span>
<span><span class="va">fragObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/Fragments.html" class="external-link">Fragments</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<p>Signac’s training dataset only has one sample, so to simulate
multiple samples, we will duplicate the data here. <strong><em>This
should not be necessary with a large dataset.</em></strong></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmc</span>, </span>
<span>  y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">pbmc</span><span class="op">)</span>, </span>
<span>  add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Sample1'</span>, <span class="st">'Sample2'</span>, <span class="st">'Sample3'</span><span class="op">)</span>,</span>
<span>  project <span class="op">=</span> <span class="st">'DuplicateData'</span></span>
<span><span class="op">)</span> </span></code></pre></div>
<p>For a larger dataset, you would interact over the fragObj list and
extract each fragments.tsv.gz file.</p>
<p>Instead, we’re just duplicating data for the sake of this
tutorial.</p>
<p>More importantly, you do need to modify the barcode in the fragment
file so it matches the barcodes in Seurat’s metadata.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fragList</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">frags</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://stuartlab.org/signac/reference/GetFragmentData.html" class="external-link">GetFragmentData</a></span><span class="op">(</span><span class="va">fragObj</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">frags</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="st">'start'</span>, <span class="st">'end'</span>, <span class="st">'barcode'</span>, <span class="st">'val'</span><span class="op">)</span></span>
<span>    <span class="va">frags</span><span class="op">$</span><span class="va">barcode</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Sample"</span>,<span class="va">x</span>,<span class="st">"_"</span>, <span class="va">frags</span><span class="op">$</span><span class="va">barcode</span>, sep <span class="op">=</span><span class="st">''</span><span class="op">)</span></span>
<span>    <span class="va">frags</span></span>
<span><span class="op">}</span>, mc.cores <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fragList</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Sample1'</span>, <span class="st">'Sample2'</span>, <span class="st">'Sample3'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="format-metadata-and-fragments-for-mocha">Format metadata and fragments for MOCHA<a class="anchor" aria-label="anchor" href="#format-metadata-and-fragments-for-mocha"></a>
</h3>
<p>Generate your sample/cell type list by finding all combinations of
Samples and Cell Populations</p>
<p>Here we must also rename the GRangesList to have names in the format
<code>CellPopulation#Sample</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">celltype_sample_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">predicted.id</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fragList</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">paste</span>, collapse <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></span>
<span><span class="co"># Extract metadata, add the Sample column, as well as CellBarcode. </span></span>
<span><span class="va">fullMeta</span> <span class="op">&lt;-</span> <span class="va">full_pbmc</span><span class="op">@</span><span class="va">meta.data</span></span>
<span><span class="va">fullMeta</span><span class="op">$</span><span class="va">Sample</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_.*"</span>,<span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">full_pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fullMeta</span><span class="op">$</span><span class="va">CellBarcode</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>,<span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">full_pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change the CellPopulation_Sample format to CellPopulation#Sample for MOCHA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fullMeta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>,<span class="st">"#"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fullMeta</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">CellType_GRanges</span> <span class="op">&lt;-</span> <span class="fu">pbapply</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pbapply/man/pbapply.html" class="external-link">pblapply</a></span><span class="op">(</span>cl <span class="op">=</span> <span class="fl">30</span>, <span class="va">celltype_sample_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"#.*"</span>,<span class="st">""</span>, <span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*#"</span>,<span class="st">""</span>, <span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">sortedMeta</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fullMeta</span>, <span class="va">predicted.id</span> <span class="op">==</span> <span class="va">celltype</span>, <span class="va">Sample</span> <span class="op">==</span> <span class="va">sample</span><span class="op">)</span></span>
<span>  <span class="va">sortedFrags</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fragList</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span>, <span class="va">barcode</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sortedMeta</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span><span class="va">sortedFrags</span>, keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">CellType_GRanges</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">celltype_sample_list</span></span></code></pre></div>
<p>Calculate the study signal (the median number of fragments per
cell)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">avg_reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fragList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>                <span class="va">filtFrag</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="va">barcode</span>  <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fullMeta</span><span class="op">)</span><span class="op">)</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">filtFrag</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">studySignal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">avg_reads</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="call-open-tiles-with-mocha">Call Open Tiles with MOCHA<a class="anchor" aria-label="anchor" href="#call-open-tiles-with-mocha"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Our blacklist comes included with Signac</span></span>
<span><span class="va">blackList</span> <span class="op">&lt;-</span> <span class="va">blacklist_hg38_unified</span></span>
<span><span class="co"># Call Open Tiles</span></span>
<span><span class="va">tileResults</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/callOpenTiles-methods.html">callOpenTiles</a></span><span class="op">(</span>ATACFragments <span class="op">=</span> <span class="va">CellType_GRanges</span>,</span>
<span>                             cellColData <span class="op">=</span> <span class="va">fullMeta</span>, </span>
<span>                             blackList <span class="op">=</span> <span class="va">blackList</span>, </span>
<span>                             genome <span class="op">=</span> <span class="st">'BSgenome.Hsapiens.UCSC.hg38'</span>,</span>
<span>                             cellPopLabel <span class="op">=</span> <span class="st">'predicted.id'</span>,</span>
<span>                             cellPopulations <span class="op">=</span> <span class="va">fullMeta</span><span class="op">$</span><span class="va">predicted.id</span>,</span>
<span>                             studySignal <span class="op">=</span> <span class="va">studySignal</span>,</span>
<span>                             cellCol <span class="op">=</span> <span class="st">'barcode'</span>,</span>
<span>                             TxDb <span class="op">=</span> <span class="st">"TxDb.Hsapiens.UCSC.hg38.refGene"</span>,</span>
<span>                             Org <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>, </span>
<span>                            outDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">'/MOCHA_Out'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span>, numCores<span class="op">=</span> <span class="fl">35</span><span class="op">)</span></span>
<span></span>
<span><span class="va">TSAM</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/getSampleTileMatrix.html">getSampleTileMatrix</a></span><span class="op">(</span><span class="va">tileResults</span>, threshold <span class="op">=</span> <span class="fl">0.2</span>, numCores <span class="op">=</span> <span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="snapatac">Importing from SnapATAC<a class="anchor" aria-label="anchor" href="#snapatac"></a>
</h2>
<p>In this example, we follow the <a href="https://github.com/r3fang/SnapATAC/tree/master/examples/10X_PBMC_15K#data_download" class="external-link">SnapATAC
10X PBMC tutorial</a> through the clustering step before extracting the
fragments and cell metadata necessary for MOCHA.</p>
<p>If you have a fully-formed Snap file for your analysis with
clustering results and sample information added to the metadata, skip to
section <a href="#format-metadata">Formatting Snap File Metadata</a>. If
following the vignette, we STRONGLY recommending installing this patched
version of SnapATAC from <a href="https://github.com/imran-aifi/SnapATAC" class="external-link">this repository</a> with
<code>devtools::install_github("imran-aifi/SnapATAC")</code>.</p>
<div class="section level3">
<h3 id="snapatac-tutorial-through-clustering">SnapATAC Tutorial through Clustering<a class="anchor" aria-label="anchor" href="#snapatac-tutorial-through-clustering"></a>
</h3>
<p>Download the all the SnapATAC Tutorial Data (linked above) to your
working directory, and load it:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CLI tool for installing from Google Drive share links</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install gdown</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1YiYd_Ydes3tqsJGpNuqQquUOoVj2EEjE/view?usp=share_link</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1YiYd_Ydes3tqsJGpNuqQquUOoVj2EEjE</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1NvGn4M2_HD06PL5Nj2if5xO-uVA0y8Q5/view?usp=share_link</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1NvGn4M2_HD06PL5Nj2if5xO-uVA0y8Q5</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1LUOqsXoQN6lVx-4RNlgH90e5oXQ0y9Bd/view?usp=share_link</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1LUOqsXoQN6lVx-4RNlgH90e5oXQ0y9Bd</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1oMJ6wFsfS-q-sY_yaLEtnYebM7RrBix6/view?usp=share_link</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1oMJ6wFsfS-q-sY_yaLEtnYebM7RrBix6</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1SEFZ5CJgcmoAkmo4_1kCMYS60YOFP379/view?usp=share_link</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1SEFZ5CJgcmoAkmo4_1kCMYS60YOFP379</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co"># https://drive.google.com/file/d/1RlBvTCqz6mhaTAkfYiCdeojD-U2mN2wp/view?usp=share_link</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="ex">gdown</span> https://drive.google.com/uc<span class="pp">?</span>id=1RlBvTCqz6mhaTAkfYiCdeojD-U2mN2wp</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r3fang/SnapATAC" class="external-link">SnapATAC</a></span><span class="op">)</span>;</span>
<span><span class="va">snap.files</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"atac_pbmc_5k_nextgem.snap"</span>, </span>
<span>  <span class="st">"atac_pbmc_10k_nextgem.snap"</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">sample.names</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"PBMC 5K"</span>,</span>
<span>  <span class="st">"PBMC 10K"</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">barcode.files</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"atac_pbmc_5k_nextgem_singlecell.csv"</span>,</span>
<span>  <span class="st">"atac_pbmc_10k_nextgem_singlecell.csv"</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">x.sp.ls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">snap.files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/createSnap.html" class="external-link">createSnap</a></span><span class="op">(</span></span>
<span>      file<span class="op">=</span><span class="va">snap.files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>      sample<span class="op">=</span><span class="va">sample.names</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="op">)</span>;</span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">x.sp.ls</span><span class="op">)</span> <span class="op">=</span> <span class="va">sample.names</span>;</span>
<span><span class="va">barcode.ls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">snap.files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">barcodes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>      <span class="va">barcode.files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, </span>
<span>      head<span class="op">=</span><span class="cn">TRUE</span></span>
<span>  <span class="op">)</span>;</span>
<span>  <span class="va">barcodes</span> <span class="op">=</span> <span class="va">barcodes</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/nrow-methods.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">barcodes</span><span class="op">)</span>,<span class="op">]</span>;</span>
<span>  <span class="va">barcodes</span><span class="op">$</span><span class="va">logUMI</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">barcodes</span><span class="op">$</span><span class="va">passed_filters</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>;</span>
<span>  <span class="va">barcodes</span><span class="op">$</span><span class="va">promoter_ratio</span> <span class="op">=</span> <span class="op">(</span><span class="va">barcodes</span><span class="op">$</span><span class="va">promoter_region_fragments</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">barcodes</span><span class="op">$</span><span class="va">passed_filters</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>;</span>
<span>  <span class="va">barcodes</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">x.sp.ls</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for both datasets, we identify usable barcodes using [3.5-5] for log10(UMI) and [0.4-0.8] for promoter ratio as cutoff.</span></span>
<span><span class="va">cutoff.logUMI.low</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">3.5</span><span class="op">)</span>;</span>
<span><span class="va">cutoff.logUMI.high</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>;</span>
<span><span class="va">cutoff.FRIP.low</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.4</span><span class="op">)</span>;</span>
<span><span class="va">cutoff.FRIP.high</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.8</span><span class="op">)</span>;</span>
<span><span class="va">barcode.ls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">snap.files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">barcodes</span> <span class="op">=</span> <span class="va">barcode.ls</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>;</span>
<span>  <span class="va">idx</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>      <span class="va">barcodes</span><span class="op">$</span><span class="va">logUMI</span> <span class="op">&gt;=</span> <span class="va">cutoff.logUMI.low</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&amp;</span> </span>
<span>      <span class="va">barcodes</span><span class="op">$</span><span class="va">logUMI</span> <span class="op">&lt;=</span> <span class="va">cutoff.logUMI.high</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&amp;</span> </span>
<span>      <span class="va">barcodes</span><span class="op">$</span><span class="va">promoter_ratio</span> <span class="op">&gt;=</span> <span class="va">cutoff.FRIP.low</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&amp;</span></span>
<span>      <span class="va">barcodes</span><span class="op">$</span><span class="va">promoter_ratio</span> <span class="op">&lt;=</span> <span class="va">cutoff.FRIP.high</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="op">)</span>;</span>
<span>  <span class="va">barcodes</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span>;</span>
<span><span class="va">x.sp.ls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="va">snap.files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">barcodes</span> <span class="op">=</span> <span class="va">barcode.ls</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>;</span>
<span>  <span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp.ls</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>;</span>
<span>  <span class="va">barcode.shared</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">barcode</span>, <span class="va">barcodes</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>;</span>
<span>  <span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">barcode.shared</span>, <span class="va">x.sp</span><span class="op">@</span><span class="va">barcode</span><span class="op">)</span>,<span class="op">]</span>;</span>
<span>  <span class="va">barcodes</span> <span class="op">=</span> <span class="va">barcodes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">barcode.shared</span>, <span class="va">barcodes</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="op">]</span>;</span>
<span>  <span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span> <span class="op">=</span> <span class="va">barcodes</span>;</span>
<span>  <span class="va">x.sp</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">x.sp.ls</span><span class="op">)</span> <span class="op">=</span> <span class="va">sample.names</span>;</span>
<span><span class="va">x.sp.ls</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine two snap object</span></span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">snapRbind</span>, <span class="va">x.sp.ls</span><span class="op">)</span>;</span>
<span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">[</span><span class="st">"Sample"</span><span class="op">]</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">@</span><span class="va">sample</span>;</span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x.sp</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 2. Add cell-by-bin matrix</span></span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/addBmatToSnap.html" class="external-link">addBmatToSnap</a></span><span class="op">(</span><span class="va">x.sp</span>, bin.size<span class="op">=</span><span class="fl">5000</span><span class="op">)</span>;</span>
<span><span class="co"># Step 3. Matrix binarization</span></span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/makeBinary.html" class="external-link">makeBinary</a></span><span class="op">(</span><span class="va">x.sp</span>, mat<span class="op">=</span><span class="st">"bmat"</span><span class="op">)</span>;</span>
<span><span class="co"># Step 4. Bin filtering</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span>;</span>
<span><span class="va">black_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"hg19.blacklist.bed.gz"</span><span class="op">)</span>;</span>
<span><span class="va">black_list.gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  <span class="va">black_list</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">black_list</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">black_list</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">idy</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Hits-class.html" class="external-link">queryHits</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">feature</span>, <span class="va">black_list.gr</span><span class="op">)</span></span>
<span><span class="op">)</span>;</span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">idy</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span>,<span class="op">-</span><span class="va">idy</span>, mat<span class="op">=</span><span class="st">"bmat"</span><span class="op">]</span>;</span>
<span><span class="op">}</span>;</span>
<span><span class="va">x.sp</span></span>
<span></span>
<span><span class="co"># Remove unwanted chromosomes</span></span>
<span><span class="va">chr.exclude</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">feature</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"random|chrM"</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>;</span>
<span><span class="va">idy</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">chr.exclude</span>, collapse<span class="op">=</span><span class="st">"|"</span><span class="op">)</span>, <span class="va">x.sp</span><span class="op">@</span><span class="va">feature</span><span class="op">)</span>;</span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">idy</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span>,<span class="op">-</span><span class="va">idy</span>, mat<span class="op">=</span><span class="st">"bmat"</span><span class="op">]</span></span>
<span><span class="op">}</span>;</span>
<span><span class="va">x.sp</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The coverage of bins roughly obeys a log normal distribution. We remove the top 5% bins that overlap with invariant features such as the house keeping gene promoters.</span></span>
<span><span class="va">bin.cov</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">bmat</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>;</span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span></span>
<span>  <span class="va">bin.cov</span><span class="op">[</span><span class="va">bin.cov</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span>, </span>
<span>  xlab<span class="op">=</span><span class="st">"log10(bin cov)"</span>, </span>
<span>  main<span class="op">=</span><span class="st">"log10(Bin Cov)"</span>, </span>
<span>  col<span class="op">=</span><span class="st">"lightblue"</span>, </span>
<span>  xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">bin.cutoff</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">bin.cov</span><span class="op">[</span><span class="va">bin.cov</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span>, <span class="fl">0.95</span><span class="op">)</span>;</span>
<span><span class="va">idy</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">bin.cov</span> <span class="op">&lt;=</span> <span class="va">bin.cutoff</span> <span class="op">&amp;</span> <span class="va">bin.cov</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>;</span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span>, <span class="va">idy</span>, mat<span class="op">=</span><span class="st">"bmat"</span><span class="op">]</span>;</span>
<span><span class="va">x.sp</span></span>
<span></span>
<span><span class="co"># We will further remove any cells of bin coverage less than 1,000. The rational behind this is that some cells may have high number of unique fragments but end up with low bin coverage after filtering. This step is optional but highly recommended.</span></span>
<span><span class="va">idx</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">bmat</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1000</span><span class="op">)</span>;</span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>;</span>
<span><span class="va">x.sp</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 5. Dimensionality reduction</span></span>
<span><span class="va">row.covs.dens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html" class="external-link">density</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">[</span>,<span class="st">"logUMI"</span><span class="op">]</span>, </span>
<span>  bw <span class="op">=</span> <span class="st">'nrd'</span>, adjust <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">sampling_prob</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/approxfun.html" class="external-link">approx</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">row.covs.dens</span><span class="op">$</span><span class="va">x</span>, y <span class="op">=</span> <span class="va">row.covs.dens</span><span class="op">$</span><span class="va">y</span>, xout <span class="op">=</span> <span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">[</span>,<span class="st">"logUMI"</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">y</span> <span class="op">+</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">)</span>;</span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>;</span>
<span><span class="va">idx.landmark.ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">10000</span>, prob <span class="op">=</span> <span class="va">sampling_prob</span><span class="op">)</span><span class="op">)</span>;</span>
<span><span class="va">x.landmark.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span><span class="va">idx.landmark.ds</span>,<span class="op">]</span>;</span>
<span><span class="va">x.query.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span><span class="op">-</span><span class="va">idx.landmark.ds</span>,<span class="op">]</span>;</span>
<span><span class="va">x.landmark.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/runDiffusionMaps.html" class="external-link">runDiffusionMaps</a></span><span class="op">(</span></span>
<span>  obj<span class="op">=</span> <span class="va">x.landmark.sp</span>,</span>
<span>  input.mat<span class="op">=</span><span class="st">"bmat"</span>, </span>
<span>  num.eigs<span class="op">=</span><span class="fl">50</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">x.landmark.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">$</span><span class="va">landmark</span> <span class="op">=</span> <span class="fl">1</span>;</span>
<span><span class="va">x.query.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/runDiffusionMapsExtension.html" class="external-link">runDiffusionMapsExtension</a></span><span class="op">(</span></span>
<span>  obj1<span class="op">=</span><span class="va">x.landmark.sp</span>, </span>
<span>  obj2<span class="op">=</span><span class="va">x.query.sp</span>,</span>
<span>  input.mat<span class="op">=</span><span class="st">"bmat"</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">x.query.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">$</span><span class="va">landmark</span> <span class="op">=</span> <span class="fl">0</span>;</span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/snapRbind.html" class="external-link">snapRbind</a></span><span class="op">(</span><span class="va">x.landmark.sp</span>, <span class="va">x.query.sp</span><span class="op">)</span>;</span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">[</span><span class="st">"sample"</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>;</span>
<span><span class="va">x.sp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/runKNN.html" class="external-link">runKNN</a></span><span class="op">(</span></span>
<span>  obj<span class="op">=</span><span class="va">x.sp</span>,</span>
<span>  eigs.dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>  k<span class="op">=</span><span class="fl">15</span></span>
<span><span class="op">)</span>;</span>
<span><span class="va">x.sp</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/runCluster.html" class="external-link">runCluster</a></span><span class="op">(</span></span>
<span>  obj<span class="op">=</span><span class="va">x.sp</span>,</span>
<span>  tmp.folder<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  louvain.lib<span class="op">=</span><span class="st">"R-igraph"</span>, <span class="co">#"leiden" preferred, but may cause issues. Requires 'library(leiden)'.</span></span>
<span>  seed.use<span class="op">=</span><span class="fl">10</span>,</span>
<span>  resolution<span class="op">=</span><span class="fl">0.7</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="format-metadata">Format Snap File Metadata<a class="anchor" aria-label="anchor" href="#format-metadata"></a>
</h3>
<p>Add the computed clusters to the Snap object metadata.</p>
<p>The Snap object contains two samples, “PBMC 5K” and “PBMC 10K”. Let’s
add a “Sample” column to the metadata. Let’s also add a column “files”
pointing to the original .snap files from which each cell came.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add clusters (from SnapATAC::runCluster) to metadata</span></span>
<span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">$</span><span class="va">cluster</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">@</span><span class="va">cluster</span></span>
<span></span>
<span><span class="co"># Add Sample name to metadata (if not done previously)</span></span>
<span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">$</span><span class="va">Sample</span> <span class="op">=</span> <span class="va">x.sp</span><span class="op">@</span><span class="va">sample</span></span>
<span></span>
<span><span class="co"># Add files to metadata, indicating the original snap file each cell belongs to.</span></span>
<span><span class="va">snap.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"atac_pbmc_5k_nextgem.snap"</span>, </span>
<span>  <span class="st">"atac_pbmc_10k_nextgem.snap"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fileList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">$</span><span class="va">Sample</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="st">"PBMC 5K"</span>, <span class="va">snap.files</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">snap.files</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span><span class="op">$</span><span class="va">files</span> <span class="op">&lt;-</span> <span class="va">fileList</span></span>
<span></span>
<span><span class="co"># SAVE this metadata to disk</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">x.sp</span><span class="op">@</span><span class="va">metaData</span>, <span class="st">"./snapMetadataforMOCHA.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-fragments">Extract Fragments from Snap File<a class="anchor" aria-label="anchor" href="#extract-fragments"></a>
</h3>
<p>Now we have a Snap object with metadata containing barcodes, unique
cell ids (column cell_id), sample names, and cell populations (cluster).
We also have our HG19 blackList, <code>black_list.gr</code>.</p>
<blockquote>
<p>Note: Following the tutorial from SnapATAC can often result in a
segfault when extracting fragments with
<code><a href="https://rdrr.io/pkg/SnapATAC/man/extractReads.html" class="external-link">SnapATAC::extractReads</a></code>. We recommend running on a machine
with large RAM and avoiding parallelization.</p>
</blockquote>
<p>Next we extract reads by sample and cell population, ensuring our
final GRanges list is named following the pattern
<code>CellPopulation#Sample</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snapMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"./snapMetadataforMOCHA.csv"</span><span class="op">)</span></span>
<span><span class="va">cellCol</span> <span class="op">&lt;-</span> <span class="st">"barcode"</span></span>
<span><span class="va">cellPopLabel</span> <span class="op">&lt;-</span> <span class="st">"cluster"</span></span>
<span><span class="va">cellPopulations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">snapMetadata</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="va">allSamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">snapMetadata</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fragmentsGRangesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">allSamples</span>, <span class="kw">function</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">barcodesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cellPopulations</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cellPop</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">snapMetadata</span><span class="op">[</span><span class="va">snapMetadata</span><span class="op">$</span><span class="va">Sample</span> <span class="op">==</span> <span class="va">sample</span>,<span class="op">]</span></span>
<span>    <span class="co"># Extract barcodes for a single cell population</span></span>
<span>    <span class="va">cellPopIdx</span> <span class="op">&lt;-</span> <span class="va">snapMetadata</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="va">cellPop</span></span>
<span>    <span class="va">cellPopBarcodes</span> <span class="op">&lt;-</span> <span class="va">snapMetadata</span><span class="op">[</span><span class="va">cellPopIdx</span>,<span class="op">]</span><span class="op">$</span><span class="va">barcode</span></span>
<span>    </span>
<span>    <span class="co"># Build the file list for the selected cell barcode</span></span>
<span>    <span class="va">files</span> <span class="op">&lt;-</span> <span class="va">snapMetadata</span><span class="op">[</span><span class="va">cellPopIdx</span>,<span class="op">]</span><span class="op">$</span><span class="va">files</span></span>
<span>    </span>
<span>    <span class="co"># Extract fragments</span></span>
<span>    <span class="va">cellPopFrags</span> <span class="op">&lt;-</span> <span class="fu">SnapATAC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SnapATAC/man/extractReads.html" class="external-link">extractReads</a></span><span class="op">(</span><span class="va">cellPopBarcodes</span>, <span class="va">files</span>, do.par <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">barcodesList</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">cellPopulations</span>, <span class="va">sample</span>, sep<span class="op">=</span><span class="st">"#"</span><span class="op">)</span></span>
<span>  <span class="va">barcodesList</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Calculate the study signal (the median number of fragments per
cell)</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">avg_reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fragmentsGRangesList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">filtFrag</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="va">barcode</span>  <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">snapMetadata</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">filtFrag</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">studySignal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">avg_reads</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="call-tiles">Call Open Tiles with MOCHA<a class="anchor" aria-label="anchor" href="#call-tiles"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Call Open Tiles</span></span>
<span><span class="va">tileResults</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/callOpenTiles-methods.html">callOpenTiles</a></span><span class="op">(</span></span>
<span>  ATACFragments <span class="op">=</span> <span class="va">fragmentsGRangesList</span>,</span>
<span>  cellColData <span class="op">=</span> <span class="va">snapMetadata</span>, </span>
<span>  blackList <span class="op">=</span> <span class="va">black_list.gr</span>, </span>
<span>  genome <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>,</span>
<span>  cellPopLabel <span class="op">=</span> <span class="va">cellPopLabel</span>,</span>
<span>  cellPopulations <span class="op">=</span> <span class="va">cellPopulations</span>,</span>
<span>  studySignal <span class="op">=</span> <span class="va">studySignal</span>,</span>
<span>  cellCol <span class="op">=</span> <span class="va">cellCol</span>,</span>
<span>  TxDb <span class="op">=</span> <span class="st">"TxDb.Hsapiens.UCSC.hg38.refGene"</span>,</span>
<span>  Org <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>, </span>
<span>  outDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">'/MOCHA_Out'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>  numCores <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">TSAM</span> <span class="op">&lt;-</span> <span class="fu">MOCHA</span><span class="fu">::</span><span class="fu"><a href="../reference/getSampleTileMatrix.html">getSampleTileMatrix</a></span><span class="op">(</span><span class="va">tileResults</span>, threshold <span class="op">=</span> <span class="fl">0.2</span>, numCores <span class="op">=</span> <span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Samir Rachid Zaim, Mark-Phillip Pebworth, Imran McGrath, Lauren Okada, Xiaojun Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
