<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="getAltTSS Pulls out all peaks that fall in TSSs,
  annotates them with the name of gene, and identifies genes that have
  evidence for alternatively regulated TSSs, including both type i (only some
  of the open TSSs for a gene are significantly more (or less) accessible),
  and type ii (multiple TSSs are significant different, with some being more
  accessible and others less). Alternatively, this function  will return all
  open TSSs with differential measurements if the returnAllTSS flag is set to
  TRUE."><title>Annotate Peaks falling in Transcription Start Sites
  (TSS) and identify alternatively regulated TSSs for each gene. — getAltTSS • MOCHA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Annotate Peaks falling in Transcription Start Sites
  (TSS) and identify alternatively regulated TSSs for each gene. — getAltTSS"><meta property="og:description" content="getAltTSS Pulls out all peaks that fall in TSSs,
  annotates them with the name of gene, and identifies genes that have
  evidence for alternatively regulated TSSs, including both type i (only some
  of the open TSSs for a gene are significantly more (or less) accessible),
  and type ii (multiple TSSs are significant different, with some being more
  accessible and others less). Alternatively, this function  will return all
  open TSSs with differential measurements if the returnAllTSS flag is set to
  TRUE."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MOCHA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/COVID-walkthrough.html">Walkthrough with COVID Dataset</a>
    <a class="dropdown-item" href="../articles/ImportingFromOtherSources.html">Importing From Other Sources</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aifimmunology/MOCHA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Annotate Peaks falling in Transcription Start Sites
  (TSS) and identify alternatively regulated TSSs for each gene.</h1>
      <small class="dont-index">Source: <a href="https://github.com/aifimmunology/MOCHA/blob/HEAD/R/getAltTSS.R" class="external-link"><code>R/getAltTSS.R</code></a></small>
      <div class="d-none name"><code>getAltTSS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>getAltTSS</code> Pulls out all peaks that fall in TSSs,
  annotates them with the name of gene, and identifies genes that have
  evidence for alternatively regulated TSSs, including both type i (only some
  of the open TSSs for a gene are significantly more (or less) accessible),
  and type ii (multiple TSSs are significant different, with some being more
  accessible and others less). Alternatively, this function  will return all
  open TSSs with differential measurements if the returnAllTSS flag is set to
  TRUE.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getAltTSS</span><span class="op">(</span></span>
<span>  <span class="va">completeDAPs</span>,</span>
<span>  returnAllTSS <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nuancedTSS <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nuancedTSSGap <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  <span class="va">TxDb</span>,</span>
<span>  <span class="va">OrgDb</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>completeDAPs</dt>
<dd><p>GRanges object that contains the differential
measurements across all peaks (unfiltered DAPs). Will also work with
data.frame or data.table version of a GRanges object. If you want
alternatively regulated TSSs, the object must include a column names 'FDR',
and 'Log2FC_C', which is standard for MOCHA differentials.</p></dd>


<dt>returnAllTSS</dt>
<dd><p>Flag to return all TSSs with DAPs measurements,
without filtering for alternative TSS usage. If multiple TSSs fall within
the same tile, then that tile will be repeated for each TSS.</p></dd>


<dt>nuancedTSS</dt>
<dd><p>True/False flag to determine if alternative TSS genes
should be filtered out if all their differential TSS usage falls within too
small of a range. Default is TRUE</p></dd>


<dt>nuancedTSSGap</dt>
<dd><p>Minimum distance betweeen TSSs needed for them to
considered distinctly regulated TSSs. If two TSSs are too close, it is
unclear and highly unlikely that ATAC data can distinguish between them.
Default is 150 bp.</p></dd>


<dt>threshold</dt>
<dd><p>FDR Threshold for determining significant vs non-significant
changes in accessibility. Following MOCHA's standards, default is 0.2.</p></dd>


<dt>TxDb</dt>
<dd><p>The TxDb-class transcript annotation 
package for your organism (e.g. "TxDb.Hsapiens.UCSC.hg38.refGene"). This 
must be installed. See 
<a href="https://bioconductor.org/packages/release/data/annotation/" class="external-link">
Bioconductor AnnotationData Packages</a>.</p></dd>


<dt>OrgDb</dt>
<dd><p>The OrgDb-class genome wide annotation 
package for your organism (e.g. "org.Hs.eg.db"). This must be installed. 
See <a href="https://bioconductor.org/packages/release/data/annotation/" class="external-link">
Bioconductor AnnotationData Packages</a></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>tpeaks A GRanges containing annotated peaks falling in TSS</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Samir Rachid Zaim, Mark-Phillip Pebworth, Imran McGrath, Lauren Okada, Xiaojun Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

