<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="plotRegion Plots the region that you've summarized across
  all cell groupings (groups=initial getPopFrags() split) with optional motif
  overlay, chromosome position ideogram, and additional GRanges tracks. If
  plotting motif overlay, ensure that motif annotations have been added to
  your counts SummarizedExperiment. A basic plot can be rendered with just a
  counts SummarizedExperiment, but additional formatting arguments allow for
  further customization. Note that to show specific genes with the option
  'whichGene' the RMariaDB package must be installed."><title>plotRegion — plotRegion • MOCHA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="plotRegion — plotRegion"><meta property="og:description" content="plotRegion Plots the region that you've summarized across
  all cell groupings (groups=initial getPopFrags() split) with optional motif
  overlay, chromosome position ideogram, and additional GRanges tracks. If
  plotting motif overlay, ensure that motif annotations have been added to
  your counts SummarizedExperiment. A basic plot can be rendered with just a
  counts SummarizedExperiment, but additional formatting arguments allow for
  further customization. Note that to show specific genes with the option
  'whichGene' the RMariaDB package must be installed."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MOCHA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/COVID-walkthrough.html">COVID-walkthrough</a>
    <a class="dropdown-item" href="../articles/Comparing-Cell-Populations.html">Comparing-Cell-Populations</a>
    <a class="dropdown-item" href="../articles/Results-Tips.html">Results-Tips</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aifimmunology/MOCHA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1><code>plotRegion</code></h1>
      <small class="dont-index">Source: <a href="https://github.com/aifimmunology/MOCHA/blob/HEAD/R/plotRegion.R" class="external-link"><code>R/plotRegion.R</code></a></small>
      <div class="d-none name"><code>plotRegion.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>plotRegion</code> Plots the region that you've summarized across
  all cell groupings (groups=initial getPopFrags() split) with optional motif
  overlay, chromosome position ideogram, and additional GRanges tracks. If
  plotting motif overlay, ensure that motif annotations have been added to
  your counts SummarizedExperiment. A basic plot can be rendered with just a
  counts SummarizedExperiment, but additional formatting arguments allow for
  further customization. Note that to show specific genes with the option
  'whichGene' the <span class="pkg">RMariaDB</span> package must be installed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotRegion</span><span class="op">(</span></span>
<span>  <span class="va">countSE</span>,</span>
<span>  plotType <span class="op">=</span> <span class="st">"area"</span>,</span>
<span>  base_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  counts_color <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  range_label_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  facet_label_side <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>  counts_color_var <span class="op">=</span> <span class="st">"Groups"</span>,</span>
<span>  counts_group_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  counts_theme_ls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  motifSetName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  motif_y_space_factor <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  motif_stagger_labels_y <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  motif_weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  motif_weight_name <span class="op">=</span> <span class="st">"Motif Weight"</span>,</span>
<span>  motif_weight_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>darkblue <span class="op">=</span> <span class="op">-</span><span class="fl">10</span>, gray <span class="op">=</span> <span class="fl">0</span>, darkred <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  motif_lab_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  motif_lab_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  motif_line_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  motif_line_size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  showGene <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  whichGene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  db_id_col <span class="op">=</span> <span class="st">"REFSEQ"</span>,</span>
<span>  collapseGenes <span class="op">=</span> <span class="st">"None"</span>,</span>
<span>  gene_theme_ls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  additionalGRangesTrack <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  linkdf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  showIdeogram <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ideogram_genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  relativeHeights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Chr <span class="op">=</span> <span class="fl">0.9</span>, `Normalized Counts` <span class="op">=</span> <span class="fl">7</span>, Links <span class="op">=</span> <span class="fl">1.5</span>, Genes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    AdditionalGRanges <span class="op">=</span> <span class="fl">4.5</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>countSE</dt>
<dd><p>A SummarizedExperiment from MOCHA::getCoverage</p></dd>


<dt>plotType</dt>
<dd><p>Options include 'overlaid','area', or 'RidgePlot'. default is
'area', which will plot a separate track for each group with the area
filled in under the curve. Setting plotType to 'overlaid' will overlay
count plot histograms across samples, instead of faceting out separately.
Setting plotType to 'RidgePlot' will generate a ridgeplot across all
groups.</p></dd>


<dt>base_size</dt>
<dd><p>Numeric, default 12. Global plot base text size parameter</p></dd>


<dt>counts_color</dt>
<dd><p>Optional color palette. A named vector of color values
where names are unique values in the `color_var` column</p></dd>


<dt>range_label_size</dt>
<dd><p>Numeric value, default 4. Text size for the y-axis
range label</p></dd>


<dt>legend.position</dt>
<dd><p>Any acceptable `legend.position` argument to theme().
Default NULL will place legend for overlaid plots at (0.8,0.8), or to the
"right" for faceted plots.</p></dd>


<dt>facet_label_side</dt>
<dd><p>Direction character value, default "top". Can also be
"right", "left", or "bottom". Position of facet label.</p></dd>


<dt>counts_color_var</dt>
<dd><p>Character value, default "Groups". Column name from
countdf to use to color counts plots. Only used if counts_group_colors
provided</p></dd>


<dt>counts_group_colors</dt>
<dd><p>Optional named color vector. Values as colors,
names are levels of `counts_color_var`. If provided, will color the plots
specifically using `scale_color_manual()`</p></dd>


<dt>counts_theme_ls</dt>
<dd><p>A list of named theme arguments passed to theme(). For
example, `list(axis.ticks = element_blank())`. Default NULL will use
`.counts_plot_default_theme`.</p></dd>


<dt>motifSetName</dt>
<dd><p>The name of the motif set in ArchRProj to use for
annotation. Example: 'JasparMotifs'</p></dd>


<dt>motif_y_space_factor</dt>
<dd><p>A factor for vertical spacing between motif
labels. Default 4. Increase to make labels farther apart, decrease to make
labels closer.</p></dd>


<dt>motif_stagger_labels_y</dt>
<dd><p>= FALSE Logical value, default FALSE. If TRUE,
will  stagger motif labels in adjacent columns in the vertical direction</p></dd>


<dt>motif_weights</dt>
<dd><p>Optional numeric vector, default NULL. If provided will
be used to color motif labels by the weighted values</p></dd>


<dt>motif_weight_name</dt>
<dd><p>Character value, default "Motif Weight". Used to
label the legend for motif colors</p></dd>


<dt>motif_weight_colors</dt>
<dd><p>Named numeric vector. Names should be color values
and breaks should be the corresponding values of motif_weights. Values
outside the highest and lowest value will appear as max or min defined
color value.</p></dd>


<dt>motif_lab_size</dt>
<dd><p>Numeric value, default 1. Size of motif labels.</p></dd>


<dt>motif_lab_alpha</dt>
<dd><p>Numeric value, default 0.25. Alpha for motif labels.</p></dd>


<dt>motif_line_alpha</dt>
<dd><p>Numeric value, default 0.25. Alpha for motif lines.</p></dd>


<dt>motif_line_size</dt>
<dd><p>Numeric value, default 1. Size of motif lines.</p></dd>


<dt>showGene</dt>
<dd><p>Logical value, default TRUE. Whether or not the gene track
should be plotted.</p></dd>


<dt>whichGene</dt>
<dd><p>Name of gene for plotting this specific gene in region.</p></dd>


<dt>db_id_col</dt>
<dd><p>Character value. Column in `orgdb` containing the output id
for `whichGene` plotting. Default "REFSEQ".</p></dd>


<dt>collapseGenes</dt>
<dd><p>Options include 'collapseAll', 'longestTx', or 'None'
Default 'None' will plot the expanded view of the reference genes,
'collapseAll' if you want collapse the gene tracks into one, and
'longestTx' will only plot the longest transcript of each gene.</p></dd>


<dt>gene_theme_ls</dt>
<dd><p>Named list of parameters passed to `theme()` for the
gene plot. Default NULL will use `.gene_plot_theme`</p></dd>


<dt>additionalGRangesTrack</dt>
<dd><p>A GRanges object containing additional track
plot data</p></dd>


<dt>linkdf</dt>
<dd><p>A dataframe with co-accessible links to display as an additional
track</p></dd>


<dt>showIdeogram</dt>
<dd><p>Logical value, default TRUE. If TRUE plots the chromosome
ideogram at the top of the multi-track plot</p></dd>


<dt>ideogram_genome</dt>
<dd><p>Character value, a genome name for the ideogram plot.
Default 'hg19'.</p></dd>


<dt>relativeHeights</dt>
<dd><p>Named numeric vector of relative heights for each of
the 4 track plots to enable clean visualization when there are many tracks.
Unused tracks will be ignored. Default value = c(`Chr` = 0.9, `Normalized
Counts` = 7, `Genes`= 2, `AdditionalGRanges` = 4.5)</p></dd>


<dt>verbose</dt>
<dd><p>Set TRUE to display additional messages. Default is FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The input ggplot object with motif labels overlaid</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># my_count_SE is a counts data frame generated by extractRegion()</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simple counts + ideogram + all genes:</span></span></span>
<span class="r-in"><span><span class="fu">plotRegion</span><span class="op">(</span>countSE <span class="op">=</span> <span class="va">my_count_SE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Motif overlay for a project my_proj containing "JasparMotifs" annotations:</span></span></span>
<span class="r-in"><span><span class="fu">plotRegion</span><span class="op">(</span></span></span>
<span class="r-in"><span>  countSE <span class="op">=</span> <span class="va">my_count_SE</span>, motifSetName <span class="op">=</span> <span class="st">"JasparMotifs"</span>,</span></span>
<span class="r-in"><span>  motif_lab_alpha <span class="op">=</span> <span class="fl">1</span>, motif_line_alpha <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Motif overlay w/ weights:</span></span></span>
<span class="r-in"><span><span class="fu">plotRegion</span><span class="op">(</span></span></span>
<span class="r-in"><span>  countSE <span class="op">=</span> <span class="va">my_count_SE</span>, motifSetName <span class="op">=</span> <span class="st">"JasparMotifs"</span>, motif_lab_alpha <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  motif_line_alpha <span class="op">=</span> <span class="fl">1</span>, motif_weights <span class="op">=</span> <span class="va">my_enrichment_weights</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Samir Rachid Zaim, Mark-Phillip Pebworth, Imran McGrath, Lauren Okada, Xiaojun Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

